---
author: "Vivek.Rai"
title: "Differential Expression Analysis"
date: "8/24/2020"
output:
 html_document:
   code_folding: hide
   smart: no
   theme: lumen
   toc: yes
params:
  outdir: NULL
editor_options:
 chunk_output_type: inline
---

```{r setup, include=FALSE}
options(stringsAsFactors = F)
set.seed(2020)

suppressPackageStartupMessages({
  library(knitr)
  library(tidyverse)
  library(scales)
  library(BiocSingular)
  library(RColorBrewer)
  library(glue)
  library(tidyHeatmap)
})

theme_set(theme_custom())
```

## Load data

```{r read-metadata}
# rename Sequencing.Batch to Batch -- feedback by @steve
metadata <- readRDS(glue("{root}/work/2020-03-25_post-qc-covariates/sample_info.rds")) %>% 
  dplyr::rename(Batch = Sequencing.Batch) %>%
  dplyr::select(-starts_with("PC"), -starts_with("Time.on"), -Blood.Type, -SAM.ID, -Cause.of.Death, 
                -Age.Units, -contains("profile"), -sry_exp, -xist_exp, -Disease.Duration.if.T2D.years) %>%
  mutate_if(is.character, ~as.factor(.x))

# remove bad sample for Alpha
metadata %>% filter(Sample.ID != "MG-0153_a-ND-52")
head(metadata)
```

```{r read-counts}
counts <- readRDS(glue("{root}/work/2020-02-14_qc_summary/counts_protein-coding.rds"))
```

```{r}
celltypes <- c("Alpha.cell.RNA", "Beta.cell.RNA", "Whole.Islet.RNA")
```


## QC

**How does the number of selected genes change with different filtering criteria?**

```{r, fig.width=12, fig.height=5}
res <- do.call(rbind, lapply(celltypes, function(cell) {
  do.call(rbind, lapply(c(5, 10, 15, 20), function(i) {
    do.call(rbind,lapply(c(.1, .2, .25, .3, .4, .5), function(j) {
      data.frame(
        num_reads = i,
        sample_frac = j,
        n_genes = sum(apply(counts[, colnames(counts) %in% metadata[metadata$Type == cell, ]$Sample.ID], 1, function(x) sum(x >= i) >= (length(x) * j))),
        type = cell
      )
    }))
  }))
}))

ggplot(res, aes(num_reads, n_genes)) +  geom_line(aes(col = as.factor(sample_frac))) +
  geom_point(size = 3, shape = 21, col = 'white', fill = 'gray50') +
  geom_text(aes(label = n_genes), size = 3, nudge_y = 100) +
  labs(col = "% of samples", x = "Minimum no. of reads", y = "No. of genes") +
  facet_wrap(~type)
```

**What is the overlap between the filtered genes?**

```{r}
get_filter_gene_overlap <- function(counts) {
  count_filter_fun <- function(fn) {
    apply(counts, 1, function(x) fn(x))
  }
  
  ## ven diagram for overlap of genes
  genes_filter_df <- data.frame(
    atleast_1_all = count_filter_fun(function(x) all(x > 1)),
    atleast_5_all = count_filter_fun(function(x) all(x > 5)),
    atleast_5_50 = count_filter_fun(function(x) sum(x >= 5) >= (length(x) * .5)),
    atleast_10_50 = count_filter_fun(function(x) sum(x >= 10) >= (length(x) * .5)),
    atleast_1_cpm = count_filter_fun(function(x) mean(x) > 1)
  )
}
```


```{r, fig.height=6, fig.width=6}
par(mfrow=c(1, 3))
for (i in celltypes) {
  df <- get_filter_gene_overlap(counts[, colnames(counts) %in% dplyr::filter(metadata, Type == i)$Sample.ID])
  #writeLines(i)
  #writeLines("---------------")
  print(apply(df, 2, sum))
  #limma::vennDiagram(df[1:5], main = i)
}
```
