---
title: "Gene Network Analysis"
author: "Vivek Rai"
date: "7/8/2020"
output:
 html_document:
   code_folding: hide
   smart: no
   toc: yes
params:
  celltype: "Whole.Islet.RNA"
  power_k: -1
  min_power_threshold: 0
  group: "all"
  limma: TRUE
  outdir: "/home/vivekrai/analyses/2020-01_vanderbilt_rna/work/wgcna/2020-12-09"
  fdr: 0.01
  merge_cut_height: 0.15
editor_options:
 chunk_output_type: inline
---

```{css, echo=FALSE}
body {
  max-width: 1440px !important;
  width: 1440px !important;
}

.main-container {
  width: 1440px !important;
}
```

```{r setup, include=FALSE}
options(stringsAsFactors = F, width = 120)

knitr::opts_chunk$set(
  comment = "",
  dev = c('png', 'cairo_pdf'),
  warning = FALSE, message = FALSE, tidy = FALSE,
  size = "small",
  cache = TRUE, cache.lazy = FALSE,
  fig.width = 8, fig.height = 6, fig.keep = 'all'
)
seed <- 2020

suppressPackageStartupMessages({
  library(WGCNA)
  library(knitr)
  library(patchwork)
  library(tidyverse)
  library(glue)
})

theme_set(theme_custom())

root.dir <- "/home/vivekrai/analyses/2020-01_vanderbilt_rna"
source(glue("{root.dir}/scripts/0_utils.R"))
```

```{r, include=FALSE}
stopifnot(!is.null(params$outdir))
enableWGCNAThreads()
```

```{r global_vars}
FDR <- params$fdr
min_reads_per_sample <- 10
min_fraction_of_sample <- .25

out_dir_prefix <- glue("{params$outdir}/{params$celltype}/{params$group}_cut-{params$merge_cut_height}")
dir.create(out_dir_prefix, recursive = T, showWarnings = F)

print(params) # inspect all parameters
```

# Part 1: Setup

## Prepare data

### Sample covariate information
```{r}
metadata <- readRDS(glue("{root.dir}/work/2020-03-25_post-qc/adults-only_sample-info.rds")) %>% 
  dplyr::rename(Batch = Sequencing.Batch) %>%
  dplyr::select(-starts_with("Time.on"), -Blood.Type, -SAM.ID, -Cause.of.Death,
                -Age.Units, -contains("profile"), -sry_exp, -xist_exp) %>%
  mutate_if(is.character, ~as.factor(.x)) %>%
  mutate(Disease.Status = ifelse(grepl("Normal", Disease.Status), "Normal", "Type 2 Diabetes"))

head(metadata)
```

```{r, include=FALSE}
celltype_meta <- filter_eval(metadata, glue("Type == '{params$celltype}'")) %>%
  dplyr::select(-Type)
  
message(head(celltype_meta))
# ## Remove the bad Alpha.cell sample
# if (params$celltype %in% c("Alpha.cell.RNA")) {
#   celltype_meta <- celltype_meta %>% filter(Sample.ID != "MG-0153_a-ND-52")
# }
```

```{r}
print(glue("INFO: {nrow(celltype_meta)} samples for network construction."))
if (nrow(celltype_meta) < 15) {
  warning("Running WGCNA analysis with less than 15 samples is not recommended. Anyway..")
}
```

### Expression and covariate data
```{r counts, include=FALSE}
counts_mat <- readRDS(glue("{root.dir}/work/2020-03-25_post-qc/counts_protein-coding.rds"))
```

```{r subset}
# subset to samples
celltype_raw_counts <- counts_mat[, colnames(counts_mat) %in% celltype_meta$Sample.ID]

# reorder the metadata matrix in same order as samples in counts mat -- just to be sure
celltype_meta <- celltype_meta[match(colnames(celltype_raw_counts), celltype_meta$Sample.ID), ]
```

```{r}
# subset and remove lowly expressed genes
print("INFO: filtering genes..")
celltype_raw_counts <- subset(
  celltype_raw_counts,
  apply(celltype_raw_counts, 1, function(x) sum(x >= min_reads_per_sample) >= (length(x) * min_fraction_of_sample))
)
print(glue("INFO: kept {nrow(celltype_raw_counts)} genes."))
```

### Normalize
```{r norm}
## Method 1: normalize using varianceStabilizingTransform from DESeq2, blind = TRUE
celltype_counts <- normalize_deseq(celltype_raw_counts)

## Method 2: normalize for library sizes and log2 transform with pseudocount
# size.factors <- DESeq2::estimateSizeFactorsForMatrix(celltype_raw_counts)
# celltype_counts <- log2(t(apply(celltype_raw_counts, 1, function(x) x/size.factors)) + 1)

if (params$limma) {
  writeLines("INFO: correcting for batch effects using limma (will preserve disease status)..")
  design_mat <- as.matrix(
    celltype_meta[, c("Age", "Sex", "BMI", "TIN_mean")] %>% mutate(Sex = ifelse(Sex == "Male", 1, 0))
  )
  print("INFO: Design matrix (only few rows shown)")
  print(head(design_mat))
  
  # NOTE: limma input matrix should be log-expression values
  if (length(unique(celltype_meta$Batch)) > 1) {
    writeLines("INFO: found batch variable.")
    celltype_norm_counts <- limma::removeBatchEffect(
      celltype_counts,
      batch = celltype_meta$Batch,
      covariates = design_mat,
      design = matrix(ifelse(celltype_meta$Disease.Status == "Normal", 0, 1), nrow(celltype_meta), 1)
    )
  } else {
    writeLines("INFO: no batch info was found — will adjust for remaining covariates")
    celltype_norm_counts <- limma::removeBatchEffect(
      celltype_counts,
      covariates = design_mat,
      design = matrix(ifelse(celltype_meta$Disease.Status == "Normal", 0, 1), nrow(celltype_meta), 1)
    )
  }
}
```

```{r recode_meta_vars}
celltype_meta <- celltype_meta %>%
  dplyr::mutate(Disease.Status = dplyr::recode(
    Disease.Status,
    "Normal" = 0,
    "Type 2 Diabetes" = 1
  )) %>%
  dplyr::mutate(Sex = dplyr::recode(
    Sex,
    "Female" = 0,
    "Male" = 1
  ))
```


```{r plot_normalized_v_corrected, fig.height=4, fig.width=8}
cowplot::plot_grid(
  qplot(rowMeans(celltype_counts), geom = c("density"), xlab =  "Mean of log2 exp.", main = "Normalized"),
  qplot(rowMeans(celltype_norm_counts), geom = c("density"), xlab = "Mean of log2 exp. ", main = "Normalized + Batch corrected")
)
```

# Part 2: WGCNA

## Create network
### Check data

```{r}
# WGNCA requires samples x gene
celltype_norm_counts <- t(celltype_norm_counts)

# Use built-in function to check for good/bad genes
gsg <- goodSamplesGenes(celltype_norm_counts)
stopifnot(gsg$allOK)

saveRDS(celltype_norm_counts, file = glue("{out_dir_prefix}/celltype-norm-counts.rds"))
```

### Analysis of the network topology
```{r pick_threshold}
# Setup base parameters to be used throughout
network_type <- "signed hybrid"
cor_fun <- "bicor"

# Choose a set of soft-thresholding powers
powers <- c(c(1:10), seq(from = 11, to=20, by=1))
sft <- pickSoftThreshold(
  celltype_norm_counts,
  powerVector = powers,
  blockSize = 16000,
  networkType = network_type,
  corFnc = cor_fun
)
```

```{r fig_scale_vs_power, fig.width=6, fig.height=5}
# Scale-free topology fit index as a function of the soft-thresholding power
ggplot(data = NULL, aes(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2])) +
  labs(x ="Soft Threshold (power)", y = "Scale Free Topology Model Fit, signed R^2",
       title =  paste("Scale independence")) + geom_text(aes(label = powers), col = 'red') + 
  geom_line(alpha = .5) +
  geom_hline(aes(yintercept = .8), col = 'blue', linetype = 'dashed')
```

```{r fig_connec_vs_power, fig.width=6, fig.height=5}
# Mean connectivity as a function of the soft-thresholding power
ggplot(data = NULL, aes(sft$fitIndices[,1], sft$fitIndices[,5])) +
  labs(x = "Soft Threshold (power)", y = "Mean Connectivity", title = "Mean connectivity") +
  geom_text(aes(label = powers), col = 'red') + geom_line(alpha = .5)
```

We will use the soft-power threshold from the above figure in constructing the network
in the next step. Let's use $k$ for which the $R^2$ for fit is at least 80%.

#### What's the k?

```{r}
power_80fit <- sft$fitIndices[which(-sign(sft$fitIndices[,3])*sft$fitIndices[,2] >= .80)[1], 1]
if (power_80fit > 15) {
  warning(glue("WARN: Power value of {power_80fit} is above 15 — likely indicates subsets/batches in the data"))
}

if (params$power_k == -1) {
  power_k <- ifelse(
    power_80fit < params$min_power_threshold, params$min_power_threshold, power_80fit
  )
} else {
  power_k <- params$power_k
}

print(glue("\\
  INFO: got k = {power_80fit} for 80% fit, chose k = {power_k} \\
  for { round(-sign(sft$fitIndices[power_k, 3]) *  sft$fitIndices[power_k, 2], 4) * 100 } % fit."
  )
)
```

### Construct network using one-step method

```{r blockwise_modules}
# NOTE: There's a bug when running blockwiseModules without accounting for the namespace
# collision of `cor` function[^1]. The code below contains the fix.
cor <- WGCNA::cor

## one step method
net <- blockwiseModules(
  celltype_norm_counts, power = power_k,
  maxBlockSize = 16000,
  networkType = network_type,
  corType = cor_fun,
  numericLabels = TRUE, 
  pamRespectsDendro = FALSE,
  saveTOMs = TRUE,
  loadTOM = TRUE,
  saveTOMFileBase = glue("{out_dir_prefix}/blockwiseTOM"),
  minModuleSize = 30,
  mergeCutHeight = params$merge_cut_height,
  randomSeed = seed,
  nThreads = 3,
  verbose = 2)

cor <- stats::cor

moduleLabels <- net$colors
moduleColors <- labels2colors(moduleLabels)

saveRDS(net, file = glue("{out_dir_prefix}/blockwiseModules.rds"))
```

`net$colors` contains the module assignment, and `net$MEs` contains the eigengenes of the modules.

**How many modules are detected?**

```{r}
length(table(net$colors))
```

```{r}
summary_info <- as.data.frame(net$colors) %>%
  rownames_to_column('ensembl') %>%
  dplyr::rename(module = `net$colors`)

summary_info <- summary_info %>%
  mutate(
    symbol = get_mapping(summary_info$ensembl)$SYMBOL
  )

saveRDS(file = glue("{out_dir_prefix}/module-assignment.rds"), summary_info)
```

## Visualize network

```{r, fig.width=2, fig.height=5}
qplot('', table(net$colors), geom = c("violin", "jitter"), ylab = "Number of genes in module")
```

We will save module and eigengene data for downstream analysis. For example,
using eigengenes, we can run correlations with the external trait data to find
any associations.

**Gene correlations**

```{r calc_tom, eval=F}
TOM <- get(load(net$TOMFiles))

plot_tom <- function(tom, n_genes, sample = 500) {
  sample_idx <- sample(n_genes, size = n_genes)
  dissTOM <- 1 - as.matrix(TOM)[sample_idx, sample_idx]
  
  selectTree <- hclust(as.dist(dissTOM), method = "average")
  selectColors <- moduleColors[sample_idx]
  
  plotDiss <- dissTOM^7
  diag(plotDiss) <- NA
  
  TOMplot(
    plotDiss,
    selectTree,
    selectColors,
    main = glue("Network heatmap plot, {sample} random genes")
  )
}
```

```{r fig_tom, fig.height=10, fig.width=10, eval=F}
plot_tom(TOM, ncol(celltype_norm_counts), 500)
rm(TOM)
```


```{r fig_eigen_heatmap, fig.height=12, fig.width=7}
par(cex = .5)

plotEigengeneNetworks(
  orderMEs(net$MEs),
  "Eigengene adjacency heatmap",
  marHeatmap = c(3,4,2,2),
  plotDendrograms = TRUE,
  xLabelsAngle = 90,
  excludeGrey = TRUE
)
```

```{r fig_eigen_dendro}
plotDendroAndColors(
  net$dendrograms[[1]], moduleColors, "Module colors",
  main = "Gene dendrogram and module colors in block 1", dendroLabels = FALSE,
  hang = 0.03, addGuide = TRUE, guideHang = 0.05
)
```

## Create network adjacency and fundamental network concepts

```{r adj_mat}
message("Computing adjacency matrix..")
adj_exp <- WGCNA::adjacency(
  celltype_norm_counts,
  type = network_type,
  corFnc = cor_fun,
  power = power_k
)
saveRDS(adj_exp, file = glue("{out_dir_prefix}/adj-exp.rds"))
```

```{r net_concepts, eval=F}
message("Computing network concepts..")
network_concepts <- WGCNA::fundamentalNetworkConcepts(adj_exp)
saveRDS(network_concepts, file = glue("{out_dir_prefix}/network-concepts.rds"))
```

```{r cluster_coef, eval=F}
cluster_coef <- WGCNA::clusterCoef(adj_exp)
saveRDS(cluster_coef, file = glue("{out_dir_prefix}/cluster-coef.rds"))
```

```{r connectivity}
intramod_connect <- WGCNA::intramodularConnectivity(adj_exp, colors = net$colors) %>%
  rownames_to_column("ensembl") %>%
  mutate(symbol = get_mapping(ensembl)$SYMBOL)

saveRDS(file = glue("{out_dir_prefix}/connectivity.rds"), object = intramod_connect)
```

```{r free_mem, include=FALSE}
rm(adj_exp)
WGCNA::collectGarbage()
```

## Variance explained

```{r var_exp}
var_exp <- WGCNA::propVarExplained(
  celltype_norm_counts,
  net$colors,
  net$MEs,
  corFnc = cor_fun
)

saveRDS(file = glue("{out_dir_prefix}/var-exp.rds"), object = var_exp)
```

```{r}
plot_connectivity_vs_membership <- function(net, module_membership, connectivity, module) {
  df <- cbind(
    connectivity = connectivity[net$colors == module, 2],
    mm_score = module_membership[which(net$colors == module), which(names(net$MEs) == glue("ME{module}"))]
  ) %>%
    as.data.frame() %>%
    rownames_to_column(var = "ensembl") %>%
    mutate(symbol = get_mapping(ensembl)$SYMBOL)
  
  conn_lim <- sort(df$connectivity, decreasing = T)[10] #quantile(df$connectivity, .95)
  mm_lim <- sort(df$mm_score, decreasing = T)[10] #quantile(df$mm_score, .95)
  print(glue("INFO: Using limits c({conn_lim}, {mm_lim})"))
  
  df$zoom <- (df$connectivity > conn_lim) | (df$mm_score > mm_lim) | (df$symbol == "RFX6")
  
  ggplot(df, aes(connectivity, mm_score)) +
  geom_point(aes(col = zoom)) +
  scale_color_manual(values = c("gray", "red")) +
  ggforce::facet_zoom(xy = connectivity > conn_lim | mm_score > mm_lim) +
  ggrepel::geom_text_repel(data = df[df$zoom, ], aes(label = symbol), size = 4) +
  guides(col = F)
}
```

# Part 3

## Module-covariate associations

We'll look at the association of module eigengenes with known sample covariates such as Age, Sex, BMI, and so on. This could provide useful pointers for downstream batch correction.

```{r fun_def_plot_corr}
plot_corr <- function(cor_mat, rotate = TRUE, ...) {
  label_sig <- ifelse(sum(cor_mat$p < 0.05) > 0, "label_sig", "n")
  cmat <- cor_mat$r
  pmat <- cor_mat$p
  
  if (rotate) {
    cmat <- t(cmat)
    pmat <- t(pmat)
  }

  tryCatch({
    corrplot::corrplot(cmat, p.mat = pmat, tl.cex = .5, insig = label_sig,
                       pch.cex = 2, pch.col = "green", tl.col = "black", ...)
  }, error = function(e) {
    print("ERROR: Failed to make plot. Continuing..")
    print(e)
  })
}
```

```{r, fig.width=14, fig.height=7}
module_cov_cor <- psych::corr.test(
  net$MEs,
  DataExplorer::dummify(
    celltype_meta %>% dplyr::select(-Sample.ID, -Donor.ID, -Islet.Isolation.Center, -Race)
  ),
  use = "pairwise",
  method = "spearman",
  ci = F
)

plot_corr(module_cov_cor)
```

## Module-trait associations

We'll load the functional profiling data and combine that with sample covariate data.

```{r load_func_data}
perifusion_df <- readRDS(glue("{root.dir}/work/2020-03-23_assay-data/perifusion-20210507.rds"))

functional_df <- perifusion_df %>%
  left_join(
    celltype_meta %>% dplyr::select(Donor.ID, Sample.ID, Age, Sex, BMI, Disease.Status),
    .,
    by = c("Donor.ID" = "Donor")
  ) %>%
  dplyr::slice(match(rownames(net$MEs), Sample.ID))

print(
  glue(
    "INFO: check if rownames of MEs and functional_df are matched in order: \\
    {all(functional_df$Sample.ID == rownames(net$MEs))}"
  )
)
```

Let's compute the correlations in the next step.

```{r, fig.height=7, fig.width=14}
module_trait_cor <- psych::corr.test(
  # apply(net$MEs, 2, inv_norm),
  net$MEs,
  DataExplorer::dummify(functional_df %>% dplyr::select(-Sample.ID, -Donor.ID)),
  use = "pairwise",
  method = "spearman",
  ci = F
)

label_sig <- ifelse(sum(module_trait_cor$p < 0.05) > 0, "label_sig", "n")

corrplot::corrplot(t(module_trait_cor$r), p.mat = t(module_trait_cor$p), tl.cex = .5,
                   insig = label_sig, pch.cex = 2, pch.col = "green", tl.col = "black")
```

## Module-trait correlations, traits adjusted for covariates

Let us regress out the effects of Age, Sex, BMI from the functional profiling
data, and re-do the above plot.

```{r regress_profiles}
vars_to_corr <- colnames(perifusion_df)[!grepl("Donor", colnames(perifusion_df))]

regress_func_profile <- function(vals) {
  residuals(
    lm(inv_norm(vals) ~ functional_df$Age + functional_df$Sex + functional_df$BMI, na.action = na.exclude)
  )
}
  
regress_functional_df <- functional_df %>%
  dplyr::mutate(across(all_of(vars_to_corr), regress_func_profile))
```

```{r fig_module_trait_corr, fig.height=7, fig.width=14}
module_trait_cor <- psych::corr.test(
  net$MEs,
  regress_functional_df %>%
    dplyr::select(-Donor.ID, -Sample.ID) %>%
    DataExplorer::dummify(.),
  use = "pairwise",
  method = "spearman",
  ci = F
)

plot_corr(module_trait_cor)
```

```{r}
saveRDS(regress_functional_df, file = glue("{out_dir_prefix}/regress-functional-df.rds"))
saveRDS(module_trait_cor, file = glue("{out_dir_prefix}/module-trait-cor.rds"))
```


```{r func_def}
plot_profile_pairwise <- function(cor_obj) {
  tmp_df <- pivot_longer(
    as.data.frame(cor_obj$r) %>% rownames_to_column("module"),
    cols = -c(module),
    names_to = "profile",
    values_to = "corr"
  ) %>%
    mutate(profile = gsub("..", " ", profile, fixed = T)) %>%
    left_join(.,
    pivot_longer(
      as.data.frame(cor_obj$p) %>% rownames_to_column("module"),
      cols = -module,
      names_to = "profile",
      values_to = "pval"
      ) %>%
      mutate(profile = gsub("..", " ", profile, fixed = T)),
    by = c("module", "profile")
    )

  tidybayes::gather_pairs(tmp_df, profile, corr, triangle = "upper") %>%
    ggplot(aes(.x, .y)) +
    geom_point(size = 1) +
    geom_smooth(se = F, method = "lm") +
    geom_abline(aes(slope = 1, intercept = 0), linetype = 'dashed', col = 'gray10') +
    facet_grid(vars(.row), vars(.col), labeller = label_wrap_gen(10)) +
    labs(x = "Sperman corr. b/w profile and eigengene", y = "Spearman corr. b/w profile and eigengene", col = "Module") +
    theme(strip.text = element_text(size = 8))
}
```

### Insulin profiles

```{r fig_corr_plot_ins, fig.height=4, fig.width=12}
module_ins_cor <- psych::corr.test(
  # apply(net$MEs, 2, inv_norm),
  net$MEs,
  regress_functional_df %>% dplyr::select(matches("INS|Insulin")),
  use = "pairwise",
  method = "spearman",
  # adjust = "BY",
  ci = F
)

plot_corr(module_ins_cor)
```


```{r, eval=F, fig.width=12, fig.height=12}
plot_profile_pairwise(module_trait_cor)
```

### Glucagon profiles

```{r fig_corr_plot_gcg, fig.height=4, fig.width=12}
module_gcg_cor <- psych::corr.test(
  # apply(net$MEs, 2, inv_norm),
  net$MEs,
  regress_functional_df %>% dplyr::select(matches("GCG|Glucagon")),
  use = "pairwise",
  method = "spearman",
  # adjust = "BY",
  ci = F
)

plot_corr(module_gcg_cor)
```

```{r, eval=F, fig.width=12, fig.height=12}
plot_profile_pairwise(module_trait_cor)
```

## Module Membership and Gene Significance scores

Module membership of a gene is defined as the correlation between the expression profile
of the gene and the module eigengene. So, each gene will be correlated with all the
modules. This allows us to quantify the similarity of all genes in the matrix to every
module.

Gene significance (GS) is defined as the abs of association between the gene expression
and the trait of interest.

```{r memberships}
gene_module_membership <- WGCNA::cor(celltype_norm_counts, net$MEs, use = "pairwise.complete.obs")

gene_significance <- WGCNA::cor(
  celltype_norm_counts,
  regress_functional_df %>% dplyr::select(-Sample.ID, -Donor.ID, -Age, -Sex, -BMI),
  use = "pairwise.complete.obs",
  method = "pearson"
)

saveRDS(gene_module_membership, glue("{out_dir_prefix}/module-membership-scores.rds"))
saveRDS(gene_significance, file = glue("{out_dir_prefix}/gene-significance.rds"))
```

### `RFX6` where are you?

```{r}
genes_of_interest <- c("ENSG00000185002")
rfx6_mod <- net$colors[genes_of_interest]
rfx6_mod
```

### How many genes are in the same module with you?

```{r}
sum(net$colors == rfx6_mod)
```

### What traits are you most associated with `RFX6`?
```{r}
data.frame(sort(gene_significance[names(rfx6_mod), ]) %>% stack) %>% arrange(-abs(values))
```

### What are the top genes in the same module as measured by correlation with eigen-gene?

```{r}
rfx6_mm_df <- data.frame(
  mm_score = gene_module_membership[net$colors == rfx6_mod, paste0("ME", rfx6_mod)]
) %>%
  rownames_to_column("ensembl") %>%
  mutate(symbol = get_mapping(ensembl)$SYMBOL)

rfx6_mm_df %>%
  dplyr::arrange(-mm_score) %>% head(n = 20)
```

```{r, fig.width=12, eval=F}
mm_cutoff <- min(top_n(rfx6_mm_df, 20, mm_score)$mm_score)

ggplot(rfx6_mm_df, aes(fct_reorder(symbol, mm_score), mm_score)) +
  geom_point() +
  gghighlight::gghighlight(mm_score > mm_cutoff, label_key = symbol) +
  ggpubr::rotate_x_text(size = 0) +
  labs(x = "Genes", y = "Module membership scores") +
  ggplot(rfx6_mm_df %>% dplyr::filter(mm_score > mm_cutoff),
         aes(fct_reorder(symbol, mm_score), mm_score)) +
  geom_point() +
  ggpubr::rotate_x_text() + labs(x = "", y = "")
```


### What are the top genes w.r.t connectivity and membership in the `RFX6` containing module?

```{r fig_connec_vs_mem, fig.height=6, fig.width=10}
plot_connectivity_vs_membership(net, gene_module_membership, intramod_connect, rfx6_mod)
```

```{r fig_intraconn_hist}
ggplot(intramod_connect, aes(kWithin)) +
  geom_density(aes(y = ..density..), fill = "gray60", col = 'black', size = 1) +
  ggrepel::geom_label_repel(
    data = intramod_connect %>% dplyr::filter(symbol %in% c("RFX6", "CTCF", "TCF7L2", "PDX1")),
    aes(kWithin, 0, label = symbol),
    size = 5,
    nudge_y = .05,
    direction = "x",
    color = "red",
    angle = 90
  ) + 
  labs(x = "Within module connectivity", y = "Density") + expand_limits(x = 0, y = 0)
```

```{r fig_totalconn_hist}
ggplot(intramod_connect, aes(kTotal)) +
  geom_density(aes(y = ..density..), fill = "gray60", col = 'black', size = 1) +
  ggrepel::geom_label_repel(
    data = intramod_connect %>% dplyr::filter(symbol %in% c("RFX6", "CTCF", "TCF7L2", "PDX1")),
    aes(kTotal, 0, label = symbol),
    size = 5,
    nudge_y = .01,
    direction = "x",
    color = "red",
    angle = 90
  ) + 
  labs(x = "Total connectivity", y = "Density")
```

### What are the genes with higest connectivity to `RFX6`?

```{r, eval = F}
make_igraph <- function(TOM, net) {
  adj <- TOM
  adj[adj > 0.1] = 1
  adj[adj != 1] = 0
  network <- igraph::graph.adjacency(adj)
  network <- igraph::simplify(network) 
  V(network)$color <- net$colors
  V(network)$gene <- names(net$colors)
  network <- igraph::delete.vertices(network, degree(network) == 0)
  network
}

tmp_graph <- make_igraph(TOM, net)
plot(tmp_graph, layout=layout.fruchterman.reingold, edge.arrow.size = 0.2)
```


## GO enrichment

```{r}
go_module_res <- lapply(rfx6_mod, function(x) {
  module_num <- gsub("ME", "", x)
  
  # x is the module number
  gprofiler2::gost(names(net$colors[which(net$colors == module_num)]))
})

saveRDS(file = glue("{out_dir_prefix}/rfx6-mod_GO-summary.rds"), go_module_res)
```

### GO Term in the `RFX6` containing module

```{r fig_go_enrichment, fig.width=7}
ggplot(
  go_module_res[[1]]$result %>%
    dplyr::filter(source %in% c("KEGG", "REAC", "GO:BP", "GO:CC", "GO:MF", "WP"))
  ) +
  ggalt::geom_lollipop(
    aes(reorder(term_name, p_value), -log10(p_value), col = source), size = 1.5
  ) +
  coord_flip() +
  labs(x = "", y = "-log10(p-value)")
```

## Functional Enrichment

### KEGG
```{r go_all_kegg, fig.width=13, fig.height=16}
cp_kegg <- clusterProfiler::compareCluster(
  lapply(
    split(subset(summary_info, module != 0), subset(summary_info, module != 0)$module),
    function(x) na.omit(get_mapping(x$ensembl, column = "ENTREZID")$ENTREZID)
  ),
  fun = "enrichKEGG"
)

custom.dotplot(cp_kegg)
saveRDS(file = glue("{out_dir_prefix}/mod_KEGG-summary.rds"), cp_kegg)
```

### GOBP

```{r go_all_gobp, fig.width=13, fig.height=16}
cp_gobp <- clusterProfiler::compareCluster(
  lapply(
    split(subset(summary_info, module != 0), subset(summary_info, module != 0)$module),
    function(x) x$ensembl
  ),
  fun = "enrichGO",
  ont = "BP",
  keyType = "ENSEMBL",
  OrgDb = org.Hs.eg.db
)

simplify_cp_gobp <- clusterProfiler::simplify(
  cp_gobp,
  cutoff = 0.9,
  by = "p.adjust",
  select_fun = min,
  measure = "Wang"
)

saveRDS(file = glue("{out_dir_prefix}/mod_GOBP-summary.rds"), simplify_cp_gobp)
custom.dotplot(simplify_cp_gobp)
```

  [^1]: https://programmersought.com/article/90752004413/;jsessionid=6FFB92B74CEE61B323C7F61E9E07C364
