---
title: "Integrate RNA+WGCNA"
date: "2021-01-18"
output:
 html_document:
   code_folding: hide
   smart: no
   theme: lumen
   toc: yes
editor_options:
 chunk_output_type: inline
---

```{r setup}
suppressPackageStartupMessages({
  library(tidyverse)
  library(glue)
})

knitr::opts_chunk$set(dev = c('ragg_png', 'cairo_pdf'))

root.dir <- "/home/vivekrai/analyses/2020-01_vanderbilt_rna/"
source(glue("{root.dir}/scripts/0_utils.R"))

theme_set(theme_custom())
```

We'll perform exploratory and integrative analysis of various data modalities we
have collected so far: WGCNA, DE, ATAC-seq.

```{r functions_def}
get_module_go <- function(df, module) {
  gprofiler2::gost(df %>% filter(module == !!module) %>% pull(ensembl))
}
```

# Load data

```{r}
metadata <- readRDS(glue("{root.dir}/work/2020-03-25_post-qc/adults-only_sample-info.rds")) %>% 
  dplyr::rename(Batch = Sequencing.Batch) %>%
  dplyr::select(-starts_with("Time.on"), -Blood.Type, -SAM.ID, -Cause.of.Death,
                -Age.Units, -contains("profile"), -sry_exp, -xist_exp) %>%
  mutate_if(is.character, ~as.factor(.x)) %>%
  mutate(Disease.Status = ifelse(grepl("Normal", Disease.Status), "Normal", "Type 2 Diabetes")) %>%
  dplyr::mutate(Disease.Status = dplyr::recode(
    Disease.Status,
    "Normal" = 0,
    "Type 2 Diabetes" = 1
  )) %>%
  dplyr::mutate(Sex = dplyr::recode(
    Sex,
    "Female" = 0,
    "Male" = 1
  ))

head(metadata)
```

```{r read_gene_exp}
celltypes <- c(
  "Alpha.cell.RNA",
  "Beta.cell.RNA",
  "Whole.Islet.RNA"
)

celltypes_k <- c(4, 6, 4)
names(celltypes_k) <- celltypes

ct_data <- list()

for (i in celltypes) {
  print(glue("Reading {i}.."))
  ct_data[[i]] <- list(
    de = readRDS(glue(
      "{root.dir}/work/dge-analysis/2021-02-05_normalized-data-share/{i}-{celltypes_k[i]}.rds"
    ))$de$res,
    vst = readRDS(glue(
      "{root.dir}/work/dge-analysis/2021-02-05_normalized-data-share/{i}-{celltypes_k[i]}.rds"
    ))$de$vst,
    ruv_W = readRDS(glue(
      "{root.dir}/work/dge-analysis/2021-02-05_normalized-data-share/{i}-{celltypes_k[i]}.rds"
    ))$W,
    ruv_norm_counts = readRDS(glue(
      "{root.dir}/work/dge-analysis/2021-02-05_normalized-data-share/{i}-{celltypes_k[i]}.rds"
    ))$normCounts,
    norm_counts = readRDS(
      glue("{root.dir}/work/wgcna-explore/2021-04-07/{i}/power-80_cut-0.2/celltype-norm-counts.rds")
    ),
    wgcna = readRDS(
      glue("{root.dir}/work/wgcna-explore/2021-04-07/{i}/power-80_cut-0.2/module-assignment.rds")
    ),
    wgcna_net = readRDS(
      glue("{root.dir}/work/wgcna-explore/2021-04-07/{i}/power-80_cut-0.2/blockwiseModules.rds")
    ),
    trait_cor_old = readRDS(
      glue("{root.dir}/work/wgcna-explore/2021-04-07/{i}/power-80_cut-0.2/module-trait-cor.rds")
    ),
    go_df = readRDS(
      glue("{root.dir}/work/wgcna-explore/2021-04-07/{i}/power-80_cut-0.2/mod_GOBP-summary.rds")
    ),
    kegg_df = readRDS(
      glue("{root.dir}/work/wgcna-explore/2021-04-07/{i}/power-80_cut-0.2/mod_KEGG-summary.rds")
    ),
    connectivity = readRDS(
      glue("{root.dir}/work/wgcna-explore/2021-04-07/{i}/power-80_cut-0.2/connectivity.rds")
    ),
    mod_mem_scores = readRDS(
      glue("{root.dir}/work/wgcna-explore/2021-04-07/{i}/power-80_cut-0.2/module-membership-scores.rds")
    )
  )
}
```

Load the functional profiling data and remove transplant columns. Recompute the correlations using updated perifusion data.

```{r load_func_data, dependson='read_gene_exp'}
perifusion_df <- readRDS(glue("{root.dir}/work/2020-03-23_assay-data/perifusion-20210507.rds")) %>%
  dplyr::select(-contains("Transplant")) # remove transplant columns as they are too sparse

for (ct in celltypes) {
  net <- ct_data[[ct]]$wgcna_net
  functional_df <- perifusion_df %>% left_join(
    dplyr::select(metadata, Donor.ID, Sample.ID, Age, Sex, BMI, Disease.Status, HbA1C.pct, C.peptide.ng.mL, Disease.Duration.if.T2D.years),
    .,
    by = c("Donor.ID" = "Donor")
  ) %>%
    dplyr::slice(match(rownames(net$MEs), Sample.ID))
  
  print(
    glue(
      "INFO: check if rownames of MEs and functional_df are matched in order: \\
    {all(functional_df$Sample.ID == rownames(net$MEs))}"
    )
  )
  
  vars_to_corr <- colnames(functional_df)[!grepl("Donor|Sample|Age|Sex|BMI|Disease.Status|HbA1|Disease|peptide", colnames(functional_df))]
  
  regress_func_profile <- function(vals) {
    residuals(
      lm(inv_norm(vals) ~ functional_df$Age + functional_df$Sex + functional_df$BMI, na.action = na.exclude)
    )
  }
  
  regress_functional_df <- functional_df %>%
    dplyr::mutate(across(all_of(vars_to_corr), regress_func_profile))
  
  ct_data[[ct]]$trait_cor <- psych::corr.test(
    net$MEs,
    regress_functional_df %>% dplyr::select(-Donor.ID, -Sample.ID),
    use = "pairwise",
    method = "spearman",
    adjust = "none",
    ci = F
  )
  
  ct_data[[ct]]$regress_func_df <- regress_functional_df
}
```

# QC

## RUV-seq figures

```{r}
for (name in celltypes) {
  pdf(glue("{root.dir}/figures/draft-figs/qc/{name}-rle-pca-v1.pdf"), height = 4)
  plot_ruv_diagnostics(ct_data[[name]]$ruv_norm_counts, metadata, label = F)
  dev.off()
  
  pdf(glue("{root.dir}/figures/draft-figs/qc/{name}-rle-pca-v2.pdf"), height = 4)
  plot_ruv_diagnostics(ct_data[[name]]$ruv_norm_counts, metadata, label = T)
  dev.off()
}
```


## QoRTs figs

```{r}
all_df <- read.table("work/2020-02-14_qc_summary/all_df.txt", sep = '\t', header = T) %>%
  filter(Age >= 18) %>%
  mutate(QC = Sample.ID %in% metadata$Sample.ID)
```

```{r}
ggplot(all_df, aes(Type, gbcbias_5p_3p)) +
    geom_quasirandom(width = .25, aes(col = Disease.Status), size = 1) + geom_boxplot(width = .1, notch = T, alpha = .2, col = 'black', outlier.color = 'NA') +
    ggrepel::geom_text_repel(aes(label = ifelse(gbcbias_5p_3p < .95, Sample.ID, '')), nudge_x = .2) +
    labs(x = '', y = "5'/3' bias (high expression genes)")

save_plot(glue("{root.dir}/figures/draft-figs/qc/gbcbias.pdf"))
```

```{r, echo=F}
ggplot(all_df, aes(x = Type, y=cgd_zscore)) +
  geom_quasirandom(width = .25, aes(col = Disease.Status), size = 1) + geom_boxplot(width = .1, notch = T, alpha = .2, col = 'black', outlier.color = 'NA') +
  ggrepel::geom_text_repel(aes(label = ifelse(cgd_zscore > 2, Sample.ID, '')), nudge_x = 2) +
  labs(x = '', y = "Z-score of K-S Statistic rel. to median")

save_plot(glue("{root.dir}/figures/draft-figs/qc/cgd_zscore.pdf"))
```

```{r}
ggplot(all_df, aes(x = Type, y=ReadPairs_UniqueGene)) +
  geom_quasirandom(width = .1, aes(col = Disease.Status), size = 1) + geom_boxplot(width = .1, notch = T, alpha = .2, col = 'black', outlier.color = 'NA') +
  ggrepel::geom_text_repel(aes(label = ifelse(ReadPairs_UniqueGene < .5e7 | ReadPairs_UniqueGene > 2.9e7, Sample.ID, '')), nudge_x = 2) +
  scale_y_continuous(labels = scales::unit_format(scale = 1e-6, unit = "M")) +
  labs(x='', y = 'Read Pairs (mapping to unique genes)')

save_plot(glue("{root.dir}/figures/draft-figs/qc/read_pairs.pdf"))
```


## Correlation figs
```{r replot_latent_var_cor_alpha}
plot_latent_var_cor <- function(obj, celltype) {
  celltype_pca <- get_pca(obj$vst)
  celltype_meta <- metadata %>% dplyr::filter(Type == celltype)
  
  celltype_meta <- combine_pca_with_metadata(
    celltype_pca,
    celltype_meta %>% dplyr::select(-Islet.Isolation.Center, -Disease.Duration.if.T2D.years)
  ) # Disease.Duration.if.T2D.years leads to NA in correlation
  
 combine_by_sampleid(
    celltype_meta,
    data.frame(obj$ruv_W) %>% tibble::rownames_to_column("Sample.ID")
  )
}
```

```{r}
ht_cor_alpha <- ComplexHeatmap::Heatmap(get_correlation(plot_latent_var_cor(ct_data$Alpha.cell.RNA, "Alpha.cell.RNA"), method = "pearson")$r, border = T, name = "Alpha")
tidyHeatmap::save_pdf(ht_cor_alpha, filename = glue("{root.dir}/figures/draft-figs/qc/alpha-latent-cor.pdf"), height = 8, width = 8.5)

ht_cor_beta <- ComplexHeatmap::Heatmap(get_correlation(plot_latent_var_cor(ct_data$Beta.cell.RNA, "Beta.cell.RNA"), method = "pearson")$r, border = T, name = "Beta")
tidyHeatmap::save_pdf(ht_cor_beta, filename = glue("{root.dir}/figures/draft-figs/qc/beta-latent-cor.pdf"), height = 8, width = 8.5)

ht_cor_islet <- ComplexHeatmap::Heatmap(get_correlation(plot_latent_var_cor(ct_data$Whole.Islet.RNA, "Whole.Islet.RNA"), method = "pearson")$r, border = T, name = "Islet")
tidyHeatmap::save_pdf(ht_cor_islet, filename = glue("{root.dir}/figures/draft-figs/qc/islet-latent-cor.pdf"), height = 8, width = 8.5)
```


# Differential Expression

```{r replot_dge_volcano, fig.height=8, fig.width=8}
alpha_highlight_genes <- c("HJURP", "RFX6","IER3", "CDCA2", "KCNK16", "LDLR", "KLF11", "GCG", "PPY", "CALB1", "SPOCK1", "CAPSL", "IL1RAP", "CFAP54", "FGP", "LAMA1", "PCSK1")

p_alpha <- EnhancedVolcano::EnhancedVolcano(ct_data$Alpha.cell.RNA$de,
  lab = get_mapping(rownames(ct_data$Alpha.cell.RNA$de))$SYMBOL,
  x = 'log2FoldChange',
  y = 'padj',
  selectLab = alpha_highlight_genes,
  encircle = alpha_highlight_genes,
  encircleCol = "black",
  title = 'Alpha (T2D vs ND)',
  pCutoff = 0.01,
  FCcutoff = 0.585,
  drawConnectors = T,
  pointSize = 2.0,
  widthConnectors = 0.75,
  col = c("grey70", "grey70", "grey70", "red"),
  labSize = 4)

save_plot(glue("{root.dir}/figures/draft-figs/de/alpha_volcano"), plot = p_alpha, base_width = 7, base_height = 8)
p_alpha
```

```{r, fig.height=8, fig.width=8}
beta_highlight_genes <- c("RFX6", "NKX6-1", "PAX6", "MMUT", "CFAP43", "BEST4", "RAMP3", "FAIM2", "TRPC6", "BARX1", "FLT4", "IAPP", "G6PC2", "COL3A1", "CD82", "OMA1", "GLP1R")

p_beta <- EnhancedVolcano::EnhancedVolcano(ct_data$Beta.cell.RNA$de,
  lab = get_mapping(rownames(ct_data$Beta.cell.RNA$de))$SYMBOL,
  x = 'log2FoldChange',
  y = 'padj',
  selectLab = beta_highlight_genes,
  encircle = beta_highlight_genes,
  encircleCol = "black",
  title = 'Beta (T2D vs ND)',
  pCutoff = 0.01,
  FCcutoff = 0.585,
  drawConnectors = T,
  pointSize = 2.0,
  widthConnectors = 0.75,
  col = c("grey70", "grey70", "grey70", "green"),
  labSize = 5)

save_plot(glue("{root.dir}/figures/draft-figs/de/beta_volcano"), plot = p_beta, base_width = 7, base_height = 8)
p_beta
```

```{r, fig.height=8, fig.width=8}
islet_highlight_genes <- c("INS", "GCG", "SST", "ISL1", "KCNJ8", "SSTR5", "PCOLCE2", "GALNT9", "SLITRK6", "SOX5", "PTP4A3", "DOT1L", "IL1B", "NEUROD1", "CXCL8", "ADAMTS4", "GABRA2", "FUT5", "EREG", "IL5RA", "PLIN2")

p_islet <- EnhancedVolcano::EnhancedVolcano(ct_data$Whole.Islet.RNA$de,
  lab = get_mapping(rownames(ct_data$Whole.Islet.RNA$de))$SYMBOL,
  x = 'log2FoldChange',
  y = 'padj',
  selectLab = islet_highlight_genes,
  encircle =  islet_highlight_genes,
  encircleCol = "black",
  title = 'Whole Islet (T2D vs ND)',
  pCutoff = 0.01,
  FCcutoff = 0.585,
  drawConnectors = T,
  pointSize = 2.0,
  widthConnectors = 0.75,
  col = c("grey70", "grey70", "grey70", "blue"),
  labSize = 5)

save_plot(glue("{root.dir}/figures/draft-figs/de/islet_volcano"), plot = p_islet, base_width = 7, base_height = 8)
p_islet
```

# WGCNA

```{r}
genes_of_interest <- ct_data$Alpha.cell.RNA$wgcna %>% 
  filter(grepl("^RFX\\d+", symbol)) %>% dplyr::select(ensembl, symbol)
```


## Compare old vs new correlation values

```{r}
plot_compare_old_vs_new_cor <- function(old_cor, new_cor, label, highlight.module = "ME0") {
  colnames(old_cor) <- c(
    "Age", "BMI", "Insulin Content", "Basal Insulin",
    "INS AUC: 16.7 G", "INS AUC 16.7 G 1st", "INS AUC 16.7 G 2nd",
    "INS AUC: 16.7 G + IBMX", "INS AUC: 1.7 G + Epi",
    "INS AUC: KCl", "Glucagon Content", "Basal glucagon",
    "GCG AUC: 16.7 G", "GCG AUC: 16.7 G + IBMX",
    "GCG AUC: 1.7 G + Epi", "GCG AUC: KCl",
    "Transplant: 0' Hu INS", "Transplant 0' Hu INS glucose",
    "Transplant: 15' Hu INS",
    "Transplant: 15' Hu INS glucose",
    "Sex_Female",
    "Sex_Male",
    "Disease.Status_Normal",
    "Disease.Status_Type.2.Diabetes"
  )
  
  old_long <- old_cor %>% as.data.frame %>% 
    dplyr::select(-Disease.Status_Normal, -Sex_Male) %>%
    dplyr::rename(
      Disease.Status = Disease.Status_Type.2.Diabetes,
      Sex = Sex_Female
    ) %>%
    rownames_to_column("module") %>%
    pivot_longer(cols = -module, values_to = "old_cor")
  
  new_long <- new_cor %>% as.data.frame %>% 
    rownames_to_column("module") %>%
    dplyr::select(-Age, -Sex, -BMI, -Disease.Status) %>%
    pivot_longer(cols = -module, values_to = "new_cor")
  
  comb_df <- inner_join(new_long, old_long)
  comb_df$highlight <- comb_df$module == highlight.module
  
  ggplot(comb_df, aes(new_cor, old_cor)) +
    geom_point(aes(col = highlight)) +
    geom_smooth(method = "lm") +
    scale_color_manual(values = c("gray", "red")) +
    facet_wrap(~name) +
    labs(x = "Old correlation values (Eigengene vs trait)", y = "New correlation values (Eigengene vs trait)", col = "RFX6 module", title = label)
}
```

```{r}
plot_compare_old_vs_new_cor(
  ct_data$Beta.cell.RNA$trait_cor_old,
  ct_data$Beta.cell.RNA$trait_cor,
  "Beta.cell.RNA", "ME1"
)
```

```{r, eval=F}
# Function used to compare Spearman/Pearson correlation values to see if they make a difference
compare_two_cors <- function(old_cor, new_cor, label, highlight.module = "ME0") {
    old_long <- old_cor %>% as.data.frame %>% 
    rownames_to_column("module") %>%
    pivot_longer(cols = -module, values_to = "old_cor")
  
  new_long <- new_cor %>% as.data.frame %>% 
    rownames_to_column("module") %>%
    pivot_longer(cols = -module, values_to = "new_cor")
  
  comb_df <- inner_join(new_long, old_long)
  comb_df$highlight <- comb_df$module == highlight.module
  
  ggplot(comb_df, aes(new_cor, old_cor)) +
    geom_point(aes(col = highlight)) +
    geom_smooth(method = "lm") +
    geom_abline(aes(intercept = 0, slope = 1), linetype = 'dashed') +
    scale_color_manual(values = c("gray", "red")) +
    facet_wrap(~name) +
    labs(x = "Old correlation values", y = "New correlation values", col = "", title = label)
}
```

## Enrichment of DE/Cilia genes in modules

```{r load_external_data}
cilia_genes <- read.table(
  glue("{root.dir}/data/external/2021-02-05_CiliaCarta.csv"),
  sep = ",", quote = c('"'), header = T
)
```

Given (a) list of "genes", and (b) results from WGCNA containing genes grouped
into modules, compute the enrichment of (a) in (b). Use Fisher's two-sided exact
test to quantify this.

**Question** Check if the differential genes are enriched in any specific module?

```{r}
helper_calc_enrichment_de <- function(mod_res, name, de, fdr = 0.01) {
  all_genes <- unique(c(mod_res$ensembl, rownames(de)))
  de_or_not <- all_genes %in% rownames(subset(de, padj < 0.01))
  
  mod_res %>% group_by(module) %>%
    dplyr::group_map(function(.x, .y) {
      module_or_not <- all_genes %in% .x$ensembl
      test <- fisher.test(de_or_not, module_or_not)
      .y$p.value <- test$p.value
      .y$estimate <- test$estimate
      .y
    }) %>%
    dplyr::bind_rows() %>%
    dplyr::mutate(padj = p.adjust(p.value, method = "holm")) %>%
    dplyr::mutate(type = name)
}

fisher_tests <- do.call(rbind, lapply(names(ct_data), function(name) {
  x <- ct_data[[name]]
  helper_calc_enrichment_de(x$wgcna, name, ct_data[[name]]$de)
}))
```

```{r, fig.height=4, fig.width=8.5}
p <- ggplot(fisher_tests, aes(estimate, -log10(p.value))) +
  geom_point(aes(col = ifelse(padj < 0.01 & estimate > 1, "sig. (<0.05)", "n.s."))) +
  ggrepel::geom_text_repel(aes(
    label =  ifelse(padj < 0.01 & estimate > 1, module, "")
  )) +
  scale_color_manual(values = c("gray", "red")) +
  geom_vline(aes(xintercept = 1), linetype = "dashed") +
  labs(
    y = "-log10(p-value)", x = "Odds ratio", col = "Fisher test\n(two-sided)",
    title = "Enrichment of DE (FDR 1%) genes in modules"
  ) +
  facet_wrap(~ type, scales = "free")

save_print_plot(p, filename = glue("{root.dir}/figures/draft-figs/wgcna/de-enrich"), base_width = 10, base_height = 4)
```

```{r}
helper_calc_enrichment_cilia <- function(mod_res, name) {
  all_genes <- unique(c(cilia_genes$Ensembl.Gene.ID, mod_res$ensembl))
  cilia_or_not <- all_genes %in% cilia_genes$Ensembl.Gene.ID
  
  mod_res %>% group_by(module) %>%
    dplyr::group_map(function(.x, .y) {
      module_or_not <- all_genes %in% .x$ensembl
      test <- fisher.test(cilia_or_not, module_or_not)
      .y$p.value <- test$p.value
      .y$estimate <- test$estimate
      .y
    }) %>%
    dplyr::bind_rows() %>%
    dplyr::mutate(padj = p.adjust(p.value, method = "holm")) %>%
    dplyr::mutate(type = name)
}

fisher_tests_cilia <- do.call(rbind, lapply(names(ct_data), function(name) {
  x <- ct_data[[name]]
  helper_calc_enrichment_cilia(x$wgcna, name)
}))
```

```{r, fig.height=4, fig.width=8.5}
p <- ggplot(fisher_tests_cilia, aes(estimate, -log10(p.value))) +
  geom_point(aes(col = ifelse(padj < 0.05 & estimate > 1, "sig. (<0.05)", "n.s."))) +
  ggrepel::geom_text_repel(aes(
    label =  ifelse(padj < 0.01 & estimate > 1, module, "")
  )) +
  scale_color_manual(values = c("gray", "red")) +
  geom_vline(aes(xintercept = 1), linetype = "dashed") +
  labs(
    y = "-log10(p-value)", x = "Odds ratio", col = "Fisher test\n(two-sided)",
    title = "Enrichment of CiliaCarta genes in modules"
  ) +
  facet_wrap(~ type, scales = "free")

save_print_plot(p, filename = glue("{root.dir}/figures/draft-figs/wgcna/cilia-enrich"), base_width = 10, base_height = 4)
```


## Enrichment of DE genes in connectivity bins

```{r}
lapply(names(ct_data), function(name) {
  x <- ct_data[[name]]$connectivity
  x$bin_id <- cut_interval(x$kTotal, n = 10)
  de_genes <- rownames(subset(ct_data[[name]]$de, padj < 0.01))
  
  x <- left_join(x, ct_data[[name]]$wgcna, by = "ensembl")
  
  x %>%
    mutate(de_status = ensembl %in% de_genes) %>%
    group_by(bin_id) %>%
    summarise(
      n_obs_de_genes = sum(de_status),
      frac_genes_in_bin = n() / nrow(x),
      n_exp_de_genes = length(de_genes) * frac_genes_in_bin
    ) %>%
    mutate(enrichment_over_uniform = n_obs_de_genes / n_exp_de_genes) %>%
    mutate(bin_id = as.factor(as.numeric(bin_id)), celltype = name)
}) %>% bind_rows() %>%
  ggplot(., aes(bin_id, enrichment_over_uniform), group = 1) +
  geom_line(aes(col = celltype, group = celltype), size = 1.25) +
  geom_point() +
  geom_hline(aes(yintercept = 1), linetype = 'dashed') +
  labs(y = "Enrichment of DE genes (Obs/Exp)", x = "kTotal bins (equal width)")

save_plot(filename = glue("{root.dir}/figures/draft-figs/wgcna/de-enrich-in-connectivity-bins"), base_width = 6, base_height = 3)
```


## Connectivity figures

```{r}
plot_connectivity_vs_membership <- function(net, module_membership, connectivity, module, highlight.genes = c("RFX6")) {
  df <- cbind(
    connectivity = connectivity[net$colors == module, 2],
    mm_score = module_membership[which(net$colors == module), which(names(net$MEs) == glue("ME{module}"))]
  ) %>%
    as.data.frame() %>%
    rownames_to_column(var = "ensembl") %>%
    mutate(symbol = get_mapping(ensembl)$SYMBOL)
  
  df$highlight <- (df$symbol %in% highlight.genes)
  
  ggplot(df, aes(connectivity, mm_score)) +
  geom_point(aes(col = highlight)) +
  scale_color_manual(values = c("gray", "red")) +
  ggrepel::geom_text_repel(data = df[df$highlight, ], aes(label = symbol), size = 6) +
  guides(col = F) +
  labs(x = "Intramodular connectivity", y = "Module membership") +
  ggpubr::grids(axis = "xy")
}
```

```{r}
beta_rfx6_plot <- plot_connectivity_vs_membership(
  ct_data$Beta.cell.RNA$wgcna_net,
  ct_data$Beta.cell.RNA$mod_mem_scores,
  ct_data$Beta.cell.RNA$connectivity,
  1
)

save_plot(glue("{root.dir}/figures/draft-figs/wgcna/beta-rfx6-connectivity.pdf"), beta_rfx6_plot, base_width = 4, base_height = 5)

# export connectivity scores
connectivity_gene_list <- c("RFX6", "PDX1", "PAX4", "PAX6", "NKX6-1", "NKX2-2", "NEUROD1", "MAFB", "MAFA", "ISL1", "IRX2", "FOXA2", "HNF1A", "ARX", "CTCF", "TCF7L2", "HHEX", "SIX3", "SIX2", "FOXO1")

for (ct in celltypes) {
  write.table(
    rbind(
      ct_data[[ct]]$connectivity %>% slice_max(kWithin, n = 50),
      ct_data[[ct]]$connectivity %>% dplyr::filter(symbol %in% connectivity_gene_list)),
    file = glue("{root.dir}/figures/draft-figs/wgcna/{ct}-connectivity.tsv"),
    quote = F
  )
}
```

## Recompute module-trait correlations with updated Perifusion data (adjusted for covariates)

```{r func_plot_corr_heat}
plot_corr_heat <- function(cor_mat, rotate = TRUE, keep.cols = c(), ...) {
  cmat <- cor_mat$r
  pmat <- cor_mat$p
  
  if (!is_empty(keep.cols)) {
    cmat <- cmat[, keep.cols]
    pmat <- pmat[, keep.cols]
  }
  
  if (rotate) {
    cmat <- t(cmat)
    pmat <- t(pmat)
  }
  
  col_fun = circlize::colorRamp2(c(-1, 0, 1), c("blue", "white", "red"))
  print(range(pmat))
  
  cell_fun <- function(j, i, x, y, width, height, fill) {
    grid.rect(x = x, y = y, width = width, height = height, 
            gp = gpar(col = "gray", fill = NA))
    
    # if (pmat[i, j] < 0.05) {
    #   grid.rect(x = x, y = y, width = width, height = height, 
    #           gp = gpar(col = "gray", fill = "#dbfdce"))
    # }
    
    if (pmat[i, j] < 0.01) {
      grid.circle(x = x, y = y, r = min(unit.c(width, height)), 
                gp = gpar(fill = col_fun(cmat[i, j]), col = NA))
    } else if (pmat[i, j] < 0.05) {
      grid.circle(x = x, y = y, r = 0.75 * min(unit.c(width, height)), 
                gp = gpar(fill = col_fun(cmat[i, j]), col = NA))
    } else {
      grid.circle(x = x, y = y, r = 0.1 * min(unit.c(width, height)), 
                gp = gpar(fill = col_fun(cmat[i, j]), col = NA))
    }
    # 
    # grid.circle(x = x, y = y, r = abs(cmat[i, j]) * min(unit.c(width, height)), 
    #             gp = gpar(fill = col_fun(cmat[i, j]), col = NA))
  }

  tryCatch({
    ComplexHeatmap::Heatmap(cmat, cell_fun = cell_fun, rect_gp = gpar(type = "none"), ...)
  }, error = function(e) {
    print("ERROR: Failed to make plot. Continuing..")
    print(e)
  })
}
```


```{r fig_module_trait_corr, fig.height=7, fig.width=14}
alpha_trait_cor <- plot_corr_heat(
  ct_data$Alpha.cell.RNA$trait_cor,
  keep.cols = !grepl("INS|Insulin|Age|Sex|BMI|HbA1|Disease|peptide", colnames(ct_data$Alpha.cell.RNA$trait_cor$r)),
  column_names_gp = gpar(fontsize = 10), name = "Alpha"
)

tidyHeatmap::save_pdf(alpha_trait_cor,  glue("{root.dir}/figures/draft-figs/wgcna/alpha-trait-cor.pdf"), width = 12, height = 3)

openxlsx::write.xlsx(
  setNames(
    list(ct_data$Alpha.cell.RNA$trait_cor$r, ct_data$Alpha.cell.RNA$trait_cor$p),
    list("correlation", "pval")
  ),
  glue("{root.dir}/figures/draft-figs/wgcna/alpha-trait-cor.xlsx"),
  rowNames = T
)
```

```{r fig_module_trait_corr, fig.height=7, fig.width=14}
beta_trait_cor <- plot_corr_heat(
  ct_data$Beta.cell.RNA$trait_cor,
  keep.cols = !grepl("GCG|lucagon|Age|Sex|BMI|HbA1|Disease|peptide", colnames(ct_data$Beta.cell.RNA$trait_cor$r)),
  column_names_gp = gpar(fontsize = 10), name = "Beta"
  )

tidyHeatmap::save_pdf(beta_trait_cor,  glue("{root.dir}/figures/draft-figs/wgcna/beta-trait-cor.pdf"), width = 12, height = 3)

openxlsx::write.xlsx(
  setNames(
    list(ct_data$Beta.cell.RNA$trait_cor$r, ct_data$Beta.cell.RNA$trait_cor$p),
    list("correlation", "pval")
  ),
  glue("{root.dir}/figures/draft-figs/wgcna/beta-trait-cor.xlsx"),
  rowNames = T
)
```

```{r fig_module_trait_corr, fig.height=7, fig.width=14}
islet_trait_cor <- plot_corr_heat(
  ct_data$Whole.Islet.RNA$trait_cor,
  keep.cols = !grepl("Age|Sex|BMI|HbA1|Disease|peptide", colnames(ct_data$Whole.Islet.RNA$trait_cor$r)),
  column_names_gp = gpar(fontsize = 10), name = "Islet"
)

tidyHeatmap::save_pdf(islet_trait_cor,  glue("{root.dir}/figures/draft-figs/wgcna/islet-trait-cor.pdf"), width = 12, height = 4)

openxlsx::write.xlsx(
  setNames(
    list(ct_data$Whole.Islet.RNA$trait_cor$r, ct_data$Whole.Islet.RNA$trait_cor$p),
    list("correlation", "pval")
  ),
  glue("{root.dir}/figures/draft-figs/wgcna/islet-trait-cor.xlsx"),
  rowNames = T
)
```

```{r}
helper_calc_enrichment_generic <- function(mod_res, genelist, name) {
  entrez <- get_mapping(mod_res$ensembl, column = "ENTREZID")$ENTREZID
  all_genes <- unique(c(genelist, entrez))
  set_or_not <- all_genes %in% genelist
  
  mod_res %>% group_by(module) %>%
    dplyr::group_map(function(.x, .y) {
      module_or_not <- all_genes %in% get_mapping(.x$ensembl, column = "ENTREZID")$ENTREZID
      test <- fisher.test(set_or_not, module_or_not, alternative = "greater")
      .y$p.value <- test$p.value
      .y$estimate <- test$estimate
      .y
    }) %>%
    dplyr::bind_rows() %>%
    dplyr::mutate(padj = p.adjust(p.value, method = "holm")) %>%
    dplyr::mutate(type = name)
}
```

```{r}
plot_wgcna_eigen_summary <- function(data, mod_assign, ct.label, highlight.modules = c("ME0"), trait.cor.columns = 1:7, upper.tri = F, write_excel = FALSE, ...) {
  cell_fun <- function(j, i, x, y, w, h, col) {
    if(upper.tri & i >= j) {
      grid.rect(x, y, w, h, gp = gpar(col = "white", fill = "white"))
    }
  }
  
  tmp_mat <- cor(data$wgcna_net$MEs)
  row_gene_counts <- table(data$wgcna_net$colors)
  names(row_gene_counts) <- glue("ME{names(row_gene_counts)}")
  row_gene_counts <- row_gene_counts[match(colnames(tmp_mat), names(row_gene_counts))]
  all(names(row_gene_counts) == colnames(tmp_mat))

  
  message("Computing DE enrichments..")
  fisher_tests <- do.call(rbind, lapply(names(ct_data), function(name) {
    helper_calc_enrichment_de(data$wgcna, name, ct_data[[name]]$de)
  }))
  
  de_genes_enrich_mat <- fisher_tests %>%
    dplyr::select(module, estimate, type) %>%
    pivot_wider(names_from = type, values_from = estimate) %>%
    mutate(module = paste0("ME", module)) %>%
    column_to_rownames("module")
  
  de_genes_enrich_mat <- as.matrix(de_genes_enrich_mat[colnames(tmp_mat), ])
  
  if (write_excel)
  openxlsx::write.xlsx(
    fisher_tests,
    glue("{root.dir}/figures/draft-figs/wgcna/{ct.label}_de-genes-enrich.xlsx"),
    rowNames = T
  )
  
  message("Computing KEGG enrichments..")
  fisher_tests_broad_kegg <- do.call(
    rbind,
    lapply(names(broad_kegg_genelists), function(name) {
      helper_calc_enrichment_generic(data$wgcna, broad_kegg_genelists[[name]]$gene, name)
    }))
  
  
  if (write_excel)
  openxlsx::write.xlsx(
    fisher_tests_broad_kegg,
    glue("{root.dir}/figures/draft-figs/wgcna/{ct.label}_eigen-genelist-enrich.xlsx"),
    rowNames = T
  )
  
  fisher_tests_broad_kegg <- fisher_tests_broad_kegg %>%
    dplyr::select(module, padj, type) %>%
    dplyr::mutate(module = paste0("ME", module)) %>%
    pivot_wider(names_from = type, values_from = padj) %>%
    column_to_rownames("module")
    
  fisher_tests_broad_kegg <- fisher_tests_broad_kegg[colnames(tmp_mat), ]
  
  if (write_excel)
  openxlsx::write.xlsx(
    setNames(
      list(data$trait_cor$r[, trait.cor.columns], data$trait_cor$p[, trait.cor.columns]),
      list("correlation", "pval")
    ),
    glue("{root.dir}/figures/draft-figs/wgcna/{ct.label}_eigen-trait-cor.xlsx"),
    rowNames = T
  )
  
  broad_kegg_col_order <- seriation::seriate(
    as.matrix(fisher_tests_broad_kegg), method = "BEA")

  ha_row = rowAnnotation(
    trait_cor = data$trait_cor$r[, trait.cor.columns],
    de = de_genes_enrich_mat,
    # kegg_enrich = as.matrix(fisher_tests_broad_kegg),
    genes = anno_lines(row_gene_counts),
    mark = anno_mark(
      at = match(highlight.modules, colnames(tmp_mat)),
      labels = colnames(tmp_mat)[match(highlight.modules, colnames(tmp_mat))]
    ),
    col = list(
      trait_cor = circlize::colorRamp2(c(-1, 0, 1), c("blue", "white", "red"))
      # kegg_enrich = circlize::colorRamp2(c(0, 1), c("purple", "white"))
    ),
    border = T
  )
  
  message("Plotting..")
  ComplexHeatmap::Heatmap(
    tmp_mat,
    show_row_names = F, show_column_names = F, cluster_columns = F,
    cluster_rows = F,
    right_annotation = ha_row, border = T,
    cell_fun = cell_fun, ...
  ) +
    Heatmap(
      as.matrix(fisher_tests_broad_kegg),
      cluster_rows = F,
      column_order = unlist(broad_kegg_col_order[2]),
      col = circlize::colorRamp2(c(0, 1), c("purple", "white")),
      border = T,
      name = "kegg"
    )
}
```

```{r}
ht_beta <- plot_wgcna_eigen_summary(ct_data$Beta.cell.RNA, ct_data$Beta.cell.RNA$wgcna, "Beta.cell.RNA", upper.tri = F, highlight.modules = c("ME8", "ME1", "ME38", "ME5", "ME7", "ME6", "ME19"))

cairo_pdf(glue("{root.dir}/figures/draft-figs/wgcna/beta-compound-fisher-pval.pdf"), width = 10, height = 6)
draw(ht_beta, padding = unit(c(20, 2, 2, 2), "mm"))
dev.off()
```

```{r}
ht_alpha <- plot_wgcna_eigen_summary(ct_data$Alpha.cell.RNA, ct_data$Alpha.cell.RNA$wgcna,  "Alpha.cell.RNA", upper.tri = F, highlight.modules = c("ME3", "ME5", "ME8", "ME12", "ME16"))

cairo_pdf(glue("{root.dir}/figures/draft-figs/wgcna/alpha-compound-fisher-pval.pdf"), width = 10, height = 6)
draw(ht_alpha, padding = unit(c(20, 2, 2, 2), "mm"))
dev.off()
```

```{r}
ht_islet <- plot_wgcna_eigen_summary(ct_data$Whole.Islet.RNA, ct_data$Whole.Islet.RNA$wgcna, "Whole.Islet.RNA", upper.tri = F, highlight.modules = c("ME1", "ME2", "ME4", "ME13", "ME31", "ME49"))

cairo_pdf(glue("{root.dir}/figures/draft-figs/wgcna/islet-compound-fisher-pval.pdf"), width = 10, height = 6)
draw(ht_islet, padding = unit(c(20, 2, 2, 2), "mm"))
dev.off()
```



# Association between module gene expression and functional traits

This is alternatively also called gene membership score — correlation between
gene expression and trait correlation. The idea is that highly scoring "module
genes" and also correlated to trait would be interesting candidates.

```{r}
gene_significance <- WGCNA::cor(
  # ct_data$Beta.cell.RNA$norm_counts,
  t(apply(ct_data$Beta.cell.RNA$norm_counts, 1, inv_norm)),
  ct_data$Beta.cell.RNA$regress_func_df %>% dplyr::select(-Sample.ID, -Donor.ID, -Age, -Sex, -BMI) %>%
    mutate(across(.fns = inv_norm)),
  use = "pairwise.complete.obs",
  method = "pearson"
)
sub_df <- gene_significance[subset(ct_data$Beta.cell.RNA$wgcna, module == 6)$ensembl, ]
ComplexHeatmap::Heatmap(sub_df[, 5:12], show_row_names = F)

sub_df_long <- as.data.frame(sub_df) %>% rownames_to_column("ensembl") %>% pivot_longer(cols = -ensembl, names_to = "trait")
mod_df_long <- as.data.frame(ct_data$Beta.cell.RNA$mod_mem_scores) %>% rownames_to_column("ensembl") %>% pivot_longer(cols = -ensembl, names_to = "module") %>% filter(module == "ME6")

comb_df <- inner_join(sub_df_long, mod_df_long, by = "ensembl") %>%
  filter(!grepl("GCG|glucagon|Gluca|HbA1|Disease|C.pep|Disease", trait))

plot_list_tmp <- lapply(
  # gene_sig_long %>% group_by(trait) %>% slice_max(abs(value), n = 10) %>%
    gene_sig_long %>% group_by(module) %>% slice_max(abs(value), n = 10) %>%
    dplyr::select(-ensembl) %>% split(., .$trait),
  gridExtra::tableGrob
)

wrap_plots(plot_list_tmp)
```

```{r}
gene_sig_long <- as.data.frame(gene_significance) %>% rownames_to_column("ensembl") %>% 
  pivot_longer(cols = -ensembl, names_to = "trait", values_to = "gene_sig") %>%
  filter(!grepl("GCG|glucagon|Gluca|HbA1|Disease|C.pep|Disease", trait))

comb_df <- left_join(gene_sig_long, ct_data$Beta.cell.RNA$wgcna) %>%
  mutate(module = paste0("ME", module)) %>%
  left_join(.,
            as.data.frame(ct_data$Beta.cell.RNA$mod_mem_scores) %>%
              rownames_to_column("ensembl") %>%
              pivot_longer(cols = -ensembl, names_to = "module", values_to = "mm_score"),
            by = c("ensembl", "module")
  ) %>%
  left_join(.,
            as.data.frame(ct_data$Beta.cell.RNA$de) %>% rownames_to_column("ensembl") %>%
              dplyr::select(ensembl, log2FoldChange, pvalue, padj),
            by = "ensembl"
  )
```

```{r}
ggplot(comb_df, aes(mm_score, log2FoldChange)) +
  geom_point(col = "gray50") +
  geom_hline(aes(yintercept = 0), linetype = 'dashed', col = "black") +
  geom_point(data = subset(comb_df, module %in% c("ME1", "ME6") & padj < 0.01), aes(col = module))

```

```{r}
ggplot(comb_df, aes(gene_sig, log2FoldChange)) +
    geom_point(col = "gray50") +
  geom_point(data = subset(comb_df, module %in% c("ME1", "ME6") & padj < 0.01), aes(col = module))
  facet_wrap(~trait)
```



```{r}
ggplot(comb_df, aes(gene_sig, log2FoldChange)) +
  geom_point(col = "gray50") +
  geom_point(data = subset(comb_df, module == "ME1" & padj < 0.01), col = "red") +
  # geom_point(aes(col = ifelse(module == "ME1", T, F), alpha = ifelse(module == "ME1", T, F))) +
  # scale_color_manual(values = c("gray90", "red")) +
  # scale_alpha_manual(values = c(0.3, 0.9)) +
  # labs(col = F, alpha =F) +
  facet_wrap(~trait)
```



```{r}
ggplot(comb_df, aes(value.x, value.y, group = trait)) +
  geom_point(aes(shape = '.')) +
  geom_smooth(method = "lm") +
  facet_wrap(~trait)

ggplot(comb_df, aes(value.x, value.y, group = trait)) +
    geom_point(aes(col = ifelse(ensembl == "ENSG00000185002", T, F)), shape = '.') +
  geom_abline(aes(xintercept = ifelse(ensembl == "ENSG00000185002", value.x, 0))) +
    geom_smooth(method = "lm") +
    ggpubr::stat_cor(label.y = .05, 
             aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~"))) +
    scale_color_manual(values = c("gray", "red")) +
  labs(col = F, x = "gene significance", y = "module membership")+
    facet_wrap(~trait)

ggplot(comb_df, aes(value.x, value.y, group = trait)) +
    geom_point(aes(col = ifelse(ensembl == "ENSG00000185002", T, F))) +
    geom_vline(aes(xintercept = ifelse(ensembl == "ENSG00000185002", value.x, NA)), linetype = "dashed", col = "blue") +
    geom_smooth(method = "lm") +
    ggpubr::stat_cor(label.y = .05, 
                     aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~"))) +
    scale_color_manual(values = c("gray", "red")) +
    labs(col = F, x = "gene significance", y = "module membership")+
    facet_wrap(~trait)
```



## Functional term enrichments

### For publication figure using select terms

```{r}
gp_func_enrich <- lapply(celltypes, function(ct) {
  print(ct)
  modules <- unique(ct_data[[ct]]$wgcna$module)
  setNames(lapply(modules, function(mod) {
    gprofiler2::gost(subset(ct_data[[ct]]$wgcna, module == mod)$ensembl)
  }), modules)
})
names(gp_func_enrich) <- celltypes
```

```{r}
beta_mods <- factor(c(1, 5, 6, 7, 8, 14, 17, 38, 48), ordered=T) #c()
alpha_mods <- factor(c(2, 3, 5, 6, 7, 8, 12, 16, 41), ordered = T)
islet_mods <- factor(c( 02, 04, 13, 25, 26, 28, 32, 40, 49), ordered=T)

diane_func_terms <- list(
  "Alpha.cell.RNA" = read.table(glue("{root.dir}/data/from-diane/2021-07-21_wgcna-func-terms_alpha_v2.txt"), header = F, sep = "\t", col.names = c("Term_name", "Term_ID")),
  "Beta.cell.RNA" = read.table(glue("{root.dir}/data/from-diane/2021-07-21_wgcna-func-terms_beta_v2.txt"), header = F, sep = "\t", col.names = c("Term_name")),
  "Whole.Islet.RNA" = read.table(glue("{root.dir}/data/from-diane/2021-07-21_wgcna-func-terms_islet_v2.txt"), header = F, sep = "\t", col.names = c("Term_name", "Term_ID"))
)
```

```{r}
df_gp_func_enrich <- lapply(celltypes, function(ct) {
  modules <- names(gp_func_enrich[[ct]])
  ndf <- lapply(modules, function(mod) {
    df <- gp_func_enrich[[ct]][[mod]]$result
    df$module <- mod
    df
  }) %>% bind_rows()
  
  ndf$Type <- ct
  ndf
}) %>% bind_rows()
```

```{r}
df_gp_func_enrich <- df_gp_func_enrich %>%
  dplyr::filter(source %in% c("GO:BP", "GO:CC", "GO:MF", "REAC", "KEGG", "WP"))
```


```{r}
subset_df_gp_func <- df_gp_func_enrich %>%
  dplyr::filter(
    (Type == "Alpha.cell.RNA" & (module %in% alpha_mods)) |
    (Type == "Beta.cell.RNA" & (module %in% beta_mods)) |
    (Type == "Whole.Islet.RNA" & (module %in% islet_mods))
  )
```

```{r}
ggplot(subset_df_gp_func, aes(module, term_name)) +
  geom_point(aes(col = -log10(p_value), size = -log10(p_value))) +
  ggpubr::grids(axis = "xy") +
  facet_wrap(~Type, scales = "free_x")

save_plot(filename = glue("{root.dir}/figures/draft-figs/wgcna_select-mod_func-enrich.pdf"), base_height = 20, base_width = 12)
```

```{r}
ggplot(subset_df_gp_func %>% 
         dplyr::filter(Type == "Beta.cell.RNA" & (module %in% beta_mods)) %>%
         dplyr::filter(term_name %in% diane_func_terms$Beta.cell.RNA$Term_name) %>%
         dplyr::mutate(term_name = factor(term_name, levels = diane_func_terms$Beta.cell.RNA$Term_name, ordered = T),
                       module = factor(module, levels = beta_mods)),
       aes(module, term_name)
  ) +
  ggpubr::grids(axis = "xy") +
  geom_line(aes(x = module, y = fct_rev(term_name), group = module), col = 'gray60') +
  geom_point(aes(col = -log10(p_value), size = -log10(p_value))) +
  labs(x = "", y = "", title = "Beta cell")

save_plot(filename = glue("{root.dir}/figures/draft-figs/wgcna_beta-select-mod_func-enrich.pdf"), base_height = 9, base_width = 8)


ggplot(subset_df_gp_func %>% 
         dplyr::filter(Type == "Alpha.cell.RNA" & (module %in% alpha_mods)) %>%
         dplyr::filter(term_name %in% diane_func_terms$Alpha.cell.RNA$Term_name) %>%
         dplyr::mutate(term_name = factor(term_name, levels = diane_func_terms$Alpha.cell.RNA$Term_name, ordered = T),
                       module = factor(module, levels = alpha_mods)),
       aes(module, term_name)
  ) +
  ggpubr::grids(axis = "xy") +
  geom_line(aes(x = module, y = fct_rev(term_name), group = module), col = 'gray60') +
  geom_point(aes(col = -log10(p_value), size = -log10(p_value))) +
  labs(x = "", y = "", title = "Alpha cell")

save_plot(filename = glue("{root.dir}/figures/draft-figs/wgcna_alpha-select-mod_func-enrich.pdf"), base_height = 9, base_width = 8)


ggplot(subset_df_gp_func %>% 
         dplyr::filter(Type == "Whole.Islet.RNA" & (module %in% islet_mods)) %>%
         dplyr::filter(term_name %in% diane_func_terms$Whole.Islet.RNA$Term_name) %>%
         dplyr::mutate(term_name = factor(term_name, levels = diane_func_terms$Whole.Islet.RNA$Term_name, ordered = T),
                       module = factor(module, levels = islet_mods)),
       aes(module, term_name)
  ) +
  ggpubr::grids(axis = "xy") +
  geom_line(aes(x = module, y = fct_rev(term_name), group = module), col = 'gray60') +
  geom_point(aes(col = -log10(p_value), size = -log10(p_value))) +
  labs(x = "", y = "", title = "Whole Islet")

save_plot(filename = glue("{root.dir}/figures/draft-figs/wgcna_islet-select-mod_func-enrich.pdf"), base_height = 8, base_width = 9)
```




### Using custom lists

Vanderbilt folks want enrichment across module of select "broader KEGG terms."
For this, I'd need to download the genes in each of the category, and use the
"Universal Enrichment Analysis" approach to conduct a hypergeomtric test of
enrichment of those genes in each of the modules.

For this, we can use `keggGet("hsa03020")[[1]]$GENE` to get Gene ENTREZ IDs,
Gene Symbol, and Term Name. I'd need to create a table/mapping and then run the
enrichment.

```{r}
p <- custom.dotplot(ct_data$Alpha.cell.RNA$kegg_df, group = T, filter_pval = 0.01)
save_print_plot(p, filename = glue("{root.dir}/figures/draft-figs/wgcna/alpha-kegg-fdr01"), base_width = 10, base_height = 10)
```
```{r}
p <- custom.dotplot(ct_data$Beta.cell.RNA$kegg_df, group = T, label_format = 50, filter_pval = 0.01)
save_print_plot(p, filename = glue("{root.dir}/figures/draft-figs/wgcna/beta-kegg-fdr01"), base_width = 12, base_height = 12)
```

```{r}
p <- custom.dotplot(ct_data$Whole.Islet.RNA$kegg_df, group = T, label_format = 50, filter_pval = 0.01)
save_print_plot(p, filename = glue("{root.dir}/figures/draft-figs/wgcna/islet-kegg-fdr01"), base_width = 12, base_height = 12)
```

```{r}
broad_kegg_pathways <- list(
  "Carbohydrate metabolism" = c( 00010, 00020, 00030, 00040, 00051, 00052, 00053, 00500, 00520, 00620, 00630, 00640, 00650, 00660, 00562 ),
  "Oxidative phosphorylation" = c(00190),
  "Lipid metabolism" = c( 00061, 00062, 00071, 00072, 00073, 00100, 00120, 00121, 00140, 00561, 00564, 00565, 00600, 00590, 00591, 00592, 01040 ),
  "Amino acid metabolism" = c( 00250, 00260, 00270, 00280, 00290, 00300, 00310, 00220, 00330, 00340, 00350, 00360, 00380, 00400 ),
  "Metabolism of other amino acids" = c( 00410, 00430, 00440, 00450, 00460, 00471, 00472, 00473, 00480 ),
  "Transcription" = c( 03020, 03022, 03040 ),
  "Translation" = c( 03010, 00970, 03013, 03015, 03008 ),
  "Folding, sorting, degradation" = c( 03060, 04141, 04130, 04120, 04122, 03050, 03018 ),
  "Replication and repair" = c( 03030, 03410, 03420, 03430, 03440, 03450, 03460 ),
  "Signal transduction" = c( 02020, 04010, 04013, 04016, 04011, 04012, 04014, 04015, 04310, 04330, 04340, 04341, 04350, 04390, 04391, 04392, 04370, 04371, 04630, 04064, 04668, 04066, 04068, 04020, 04070, 04072, 04071, 04024, 04022, 04151, 04152, 04150, 04075 ),
  "Signaling molecules" = c( 04080, 04060, 04061, 04512, 04514 ),
  "Cell-growth and death" = c(
    04110, 04210, 04216, 04217, 04115, 04218
  ),
  "Cellular community" = c(
    04510, 04520, 04530, 04540, 04550
  ),
  "Immune system" = c( 04640, 04610, 04611, 04613, 04620, 04624, 04621, 04622, 04623, 04625, 04650, 04612, 04660, 04658, 04659, 04657, 04662, 04664, 04666, 04670, 04672, 04062 ),
  "Endocrine system" = c( 04911, 04910, 04922, 04923, 04920, 03320, 04929, 04912, 04913, 04915, 04914, 04917, 04921, 04926, 04935, 04918, 04919, 04928, 04916, 04924, 04614, 04925, 04927 ),
  "Endocrine and metabolic diseases" = c( 04930, 04940, 04950, 04932, 04931, 04933, 04934 )
)
```

```{r}
broad_kegg_genelists <- list()
for (category in names(broad_kegg_pathways)) {
  path_ids <- broad_kegg_pathways[[category]]
  print(glue("getting category {category}.."))
  broad_kegg_genelists[[category]] <- data.frame(term = category, gene = flatten_chr(na.omit(lapply(path_ids, function(id) {
    id <- str_pad(id, 5, "0", side = 'left')
    res <- httr::GET(glue("http://togows.dbcls.jp/entry/pathway/hsa{id}/genes.json"))
    names(unlist(httr::content(res, type = "application/json")))
  }))))
}

broad_kegg_genelists <- append(
  broad_kegg_genelists,
  list(
    "InnateDB" = data.frame(
      term = "InnateDB",
      gene = get_mapping(
        read_tsv(glue("{root.dir}/data/external/InnateDB_genes.txt"))$ensembl,
        column = "ENTREZID")$ENTREZID
    ),
    "CiliaCarta" = data.frame(
      term = "CiliaCarta",
      gene = get_mapping( cilia_genes$Ensembl.Gene.ID, column = "ENTREZID")$ENTREZID
    ),
    "Matrisome" = data.frame(
      term = "Matrisome",
      gene = get_mapping(
        read.table(glue("{root.dir}/data/external/matrisome.txt"), header = F)[[1]],
        keytype = "SYMBOL", column = "ENTREZID")$ENTREZID
    )
  )
)
```

```{r}
make_pub_go_plots <- function(obj, label, width = 12, height = 12, ...) {
  p <- custom.dotplot(obj, trim_x_labels = T, group = T, filter_pval = 0.01)
  save_print_plot(p, filename = glue("{root.dir}/figures/draft-figs/wgcna/{label}-go-fdr01"),
                  base_width = width, base_height = height)  
  
  simplify_cp <- clusterProfiler::simplify(
    obj,
    cutoff = 0.5,
    by = "p.adjust",
    select_fun = min,
    measure = "Wang"
  )
  
  p <- custom.dotplot(simplify_cp, trim_x_labels = T, group = T, filter_pval = 0.01)
  
  save_print_plot(p, filename = glue("{root.dir}/figures/draft-figs/wgcna/{label}-go-fdr01-collapsed"),
                  base_width = width, base_height = height)
}
```

```{r, echo=F}
make_pub_go_plots(ct_data$Beta.cell.RNA$go_df, "beta", height = 15, width = 12)
```

```{r, echo=F}
make_pub_go_plots(ct_data$Whole.Islet.RNA$go_df, "islet", height = 15, width = 12)
```

```{r, echo=F}
make_pub_go_plots(ct_data$Alpha.cell.RNA$go_df, "alpha", height = 14, width = 10)
```

```{r dotplot_fixes_def}
ep_str_wrap <- function(string, width) {
    x <- gregexpr(' ', string)
    vapply(seq_along(x),
           FUN = function(i) {
               y <- x[[i]]
               n <- nchar(string[i])
               len <- (c(y,n) - c(0, y)) ## length + 1
               idx <- len > width
               j <- which(!idx)
               if (length(j) && max(j) == length(len)) {
                   j <- j[-length(j)]
               }
               if (length(j)) {
                   idx[j] <- len[j] + len[j+1] > width
               }
               idx <- idx[-length(idx)] ## length - 1
               start <- c(1, y[idx] + 1)
               end <- c(y[idx] - 1, n)
               words <- substring(string[i], start, end)
               paste0(words, collapse="\n")
           },
           FUN.VALUE = character(1)
    )
}

#' default_labeller
#'
#' default labeling function that uses the
#' internal string wrapping function `ep_str_wrap`
#' @noRd
default_labeller <- function(n, ...) {
    function(str){
        str <- gsub("_", " ", str)
        stringr::str_trunc(str, n, ...)
        # ep_str_wrap(str, n)
    }
}

custom.dotplot <- function(object, x= "Cluster", colorBy="p.adjust", trim_x_labels = TRUE,
                                         showCategory=5, by="geneRatio", size="geneRatio",
                                         split=NULL, includeAll=TRUE, filter_pval = 0.05,
                                         font.size=12, title="", label_format = 50,
                                         group = FALSE, group_color = FALSE, shape = FALSE) {
    color <- NULL
    df <- fortify(object, showCategory=showCategory, by=by,
                  includeAll=includeAll, split=split)
    if (trim_x_labels) {
      df[, x] <- trimws(gsub("(\\(\\d+\\))", "", df[, x]))
      df[, x] <- factor(df[, x], levels = sort(unique(as.numeric(df[, x]))))
    }
    
    df <- subset(df, p.adjust <= filter_pval)
    
    if (by != "geneRatio")
        df$GeneRatio <- parse_ratio(df$GeneRatio)
    label_func <- default_labeller(label_format)
    if(is.function(label_format)) {
        label_func <- label_format
    }
    if (is.null(size)) size <- by
    by2 <- switch(size, rowPercentage = "Percentage", 
                        count         = "Count", 
                        geneRatio     = "GeneRatio")    
    p <- ggplot(df, aes_string(x = x, y = "Description", size = by2))      
    if (group) {
      if (group_color) {
        p <- p + geom_line(aes_string(color = "Cluster", group = "Cluster"), size=.3) + 
          ggnewscale::new_scale_colour()     
      } else {
        p <- p + geom_line(aes_string(group = "Cluster"), size=.3) + 
          ggnewscale::new_scale_colour()     
      }
    }
    
    if (shape) {
        ggstar <- "ggstar"
        require(ggstar, character.only=TRUE)
        # p <- p + ggsymbol::geom_symbol(aes_string(symbol = "Cluster", fill = colorBy)) +
        p <- p + ggstar::geom_star(aes_string(starshape="Cluster", fill=colorBy)) + 
            scale_fill_continuous(low="red", high="blue", guide=guide_colorbar(reverse=TRUE))
    }  else {
        p <- p +  geom_point(aes_string(color = colorBy)) 
    }  
    suppressMessages(print(
        p + scale_color_continuous(guide=guide_colorbar(reverse=TRUE)) +
            ylab(NULL) + ggtitle(title) + DOSE::theme_dose(font.size) +
            scale_size_continuous(range=c(3, 8)) + 
            scale_y_discrete(labels = label_func)
    ))
}
```


```{r plot_adj_network}
# adj <- readRDS(glue("{root.dir}/work/wgcna-explore/2021-04-07/Beta.cell.RNA/power-80_cut-0.2/adj-exp.rds"))
# 
# adj[adj > 0.1] = 1
# adj[adj != 1] = 0
# all_graph <- igraph::simplify(igraph::graph_from_adjacency_matrix(adj, mode = "undirected"))
# all_graph <- igraph::delete.vertices(all_graph, igraph::degree(all_graph)==0)
# all_graph <- igraph::set_vertex_attr(all_graph, "symbol", index = igraph::V(all_graph), get_mapping(igraph::V(all_graph)$name)$SYMBOL)
# 
# all_graph <- igraph::set_vertex_attr(all_graph, "degree", index = igraph::V(all_graph), igraph::V(all_graph)$degree)
# 
# igraph::degree(all_graph, "ENSG00000185002")
# 
# rfx6_graph <- igraph::induced.subgraph(
#   all_graph,
#   c("ENSG00000185002", igraph::neighbors(all_graph, "ENSG00000185002")$name)
# )
# # 
# # rfx6_graph <- igraph::set_vertex_attr(
# #   rfx6_graph,
# #   "membership_score",
# #   index = igraph::V(rfx6_graph)
# # )
# 
# # try to attach GO term information to genes
# 
# 
# #igraph::plot.igraph(rfx6_graph)
# 
# ggraph(rfx6_graph, layout = "circle") + 
#   geom_edge_hive0(edge_colour = "grey66") +
#   # geom_node_point(aes(size = degree), color = "red") +
#   geom_node_text(aes(label = symbol)) +
#   theme_graph()
```




# Integration with ATAC
## GWAS enrichment

Once the data is prepared, we will run GARFIELD to test for enrichment of trait variants
in the annotations. The result will be typically two files for each trait.

### Load data 

```{r}
assign_significance <- function(x, padj) {
  if (x >= 0.05) {
    return("Not significant")
  } else if ((x < 0.05) & (x >= padj)) {
    return("Nominally significant (P<0.05)")
  } else if (x < padj) {
    return("Significant (Bonferroni corrected\nfor effective Annots)")
  }
}

annot_groups = c(
  "hg19.tss.10kb_ext"
  # "hg19.tss.5kb_ext",
  # "hg19.tss.1kb_ext",
  # "hg19.tss.5u1d_ext"
)

trait_garf_df <- do.call(rbind, lapply(c("beta-0.2", "alpha-0.2"), function(group) {
  # NOTE: edit celltype here
  trait_enrich_files <- list.files(
    glue("{root.dir}/control/05_module-peaks-gwas/work/2021-06-09/{group}/{annot_groups[1]}/results"),
    "enrichment",
    full.names = T
  )
  
  do.call(rbind, lapply(trait_enrich_files, function(x) {
    df <- read_delim(x, delim = " ", col_types = cols())
    trait <- str_match(x, ".*\\.enrichment\\.(.*)\\.out")[2]
    mdf <- read_delim(paste0(dirname(x), "/garfield.Meff.", trait, ".out"), delim = "\t", col_types = cols(), col_names = F)
    
    df$Trait <- trait
    df$Significance <- sapply(df$Pvalue, function(x) assign_significance(x, mdf[2, 2]))
    df$PAdj <- mdf[2, 2]
    df$Group <- group
    df
  }))
}))

# NOTE: edit celltype here
d <- trait_garf_df %>%
  filter(PThresh == 5e-08) %>%
  mutate(Annotation = gsub("Module-", "", Annotation)) %>%
  filter(!grepl("T1D|proinsulin|Height|Alzhei|COVID", Trait), !grepl("Ackermann|Arda|Bulk|sciATAC", Annotation))
```

### Make it heatmap style
```{r, fig.width=12, fig.height=3}
# plot_gwas_heatmap <- function(mat, annot.thesh = 0, title = '') {
#   ggplot(mat, aes(Annotation, Trait)) +
#     geom_tile(aes(fill = ifelse(NAnnotThesh >= annot.thesh, Beta / SE, 0))) +
#     geom_point(
#       shape = ifelse(grepl("Bonferroni", mat$Significance) & (mat$NAnnotThesh >= annot.thesh), 8, 32),
#       col = "black"
#     ) +
#     ggpubr::rotate_x_text(angle = 90) +
#     scale_fill_gradient2(low = "darkblue", mid = "white", high = "red") +
#     theme(
#       text = element_text(size = 10),
#       axis.text.x = element_text(size = 8),
#       strip.text.x = element_text(size = 8),
#       panel.background = element_rect(fill = "white", colour = "black"),
#       # legend.key.size=unit(3, "mm"),
#       panel.grid = element_blank(),
#       legend.position = "right",
#       legend.text = element_text(size = 8),
#       panel.grid.major.y = element_line(colour = "grey", size = 0.2, linetype = "dashed"),
#       panel.grid.major.x = element_line(colour = "grey", size = 0.2, linetype = "dashed")
#     ) +
#     labs(x = "Annotation", fill = "Beta/SE", title = title)
# }
```

```{r}
plot_gwas_heatmap <- function(df, ...) {
  df_Z <- df %>% dplyr::select(Beta, SE, Trait, Annotation) %>% 
    transmute(Z = Beta/SE, Annotation, Trait) %>% 
    pivot_wider(names_from = Annotation, values_from = Z) %>%
    column_to_rownames("Trait") %>%
    as.matrix()
  
  df_P <- df %>% dplyr::select(Significance, Trait, Annotation) %>% 
    pivot_wider(names_from = Annotation, values_from = Significance) %>%
    column_to_rownames("Trait")
  
  cell_fun <- function(j, i, x, y, w, h, col) {
    grid.rect(x = x, y = y, width = w, height = h,  gp = gpar(col = "grey", fill = NA))
    if(grepl("Bonferroni", df_P[i, j])) {
      # grid.rect(x, y, w, h, gp = gpar(col = "white", fill = "white"))
      grid.points(x, y, pch = 8, size = unit(2, "mm"))
    }
  }
  
  ComplexHeatmap::Heatmap(df_Z, col = c("white", "red"), border = T,
                          cell_fun = cell_fun, ...)
}
```

```{r}
#plot_gwas_heatmap(beta_d, 0, title = "No filtering") +
#plot_gwas_heatmap(beta_d, 1, title = ">=1 hits") +
#plot_gwas_heatmap(beta_d, 5, title = ">=5 hits") +
#plot_gwas_heatmap(beta_d, 10, title = ">=10 hits") +
#  plot_layout(guides = 'collect')
```

```{r}
beta_d <- d %>% filter(Group == "beta-0.2") %>%
  plot_gwas_heatmap(.)

tidyHeatmap::save_pdf(beta_d, filename = glue("{root.dir}/figures/draft-figs/wgcna/beta-gwas-enrich_5e8_heatmap-v2.pdf"), width = 13, height = 2)
```

```{r}
alpha_d <- d %>% filter(Group == "alpha-0.2") %>%
  plot_gwas_heatmap(.)

tidyHeatmap::save_pdf(alpha_d, filename = glue("{root.dir}/figures/draft-figs/wgcna/alpha-gwas-enrich_5e8_heatmap-v2.pdf"), width = 13, height = 2)
```

### Make a normal plot
```{r, fig.width=10, fig.height=6}
p1 <- ggplot(d, aes(x = Annotation, y = Beta)) +
  geom_rect(aes(xmin = "1", xmax = "1", ymin = -Inf, ymax = Inf), col = "lightblue", alpha = .2) +
  geom_point(aes(colour = Significance)) +
  geom_hline(yintercept = 0, color = "black", size = 0.5) +
  geom_errorbar(data = subset(d, Pvalue < 0.05), aes(ymin = CI95_lower, ymax = CI95_upper), size = 0.2) +
  theme(
    text = element_text(size = 10),
    axis.text.x = element_text(size = 8),
    strip.text.x = element_text(size = 8),
    panel.background = element_rect(fill = "white", colour = "black"),
    legend.key.size = unit(3, "mm"),
    panel.grid = element_blank(),
    legend.position = "right",
    legend.text = element_text(size = 7),
    panel.grid.major.y = element_line(colour = "grey", size = 0.2, linetype = "dashed"),
    panel.grid.major.x = element_line(colour = "grey", size = 0.2, linetype = "dashed")
  ) +
  labs(y = "Beta (log OR)", x = "Annotations") +
  coord_flip() +
  facet_wrap(~ Trait + PThresh, nrow = 1) +
  guides(col = guide_legend(nrow = 3)) +
  scale_color_manual(values = c("red", "gray", "darkred")) +
  ylim(-5, 5)

p2 <- ggplot(d, aes(x = linkID, y = peak_count / 4)) +
  labs(y = "Number of Peaks", x = "") +
  theme(axis.text.y = element_blank()) +
  scale_y_continuous(labels = scales::label_number_auto()) +
  geom_col() +
  coord_flip() +
  theme_minimal()

p1 + p2 + plot_layout(widths = c(6, 1), guides = "collect")
save_plot(filename = "../figures/2021-02-01_garfield-t2d-enrich_beta_5e8.pdf", base_width = 12, base_height = 7)
```


## DE enrichment vs GWAS enrichment

```{r}
fisher_tests %>%
  mutate(module = as.factor(module)) %>%
  right_join(d, by = c("module" = "Annotation")) %>%
  ggplot(aes(estimate, OR)) +
  geom_point(aes(col = module == 1)) +
  ggrepel::geom_text_repel(aes(label = module)) +
  facet_wrap(~Trait, scales = "free") +
  scale_color_manual(values = c("gray50", "red")) +
  labs(x = "Enrichment of DE genes (OR)", y = "Enrichment of trait-GWAS variants (OR)") +
  guides(col = FALSE)
```

## fGWAS results

```{r}
fgwas_res <- do.call(rbind, lapply(
  list.files("../work/module-peaks_fgwas/fgwas/forward/single_annot/", "*.dat", full.names = T),
  function(x) {
    read.table(x, header = T)
  }
)) %>%
  mutate(annotation = gsub("Module-", "", annotation)) %>%
  mutate(annotation = factor(annotation, levels = unique(sort(as.numeric(annotation), decreasing = T)))) %>%
  filter(annotation != 0)
```

```{r}
ggplot(fgwas_res, aes(annotation, estimate)) +
  # geom_point(aes(colour=Significance)) +
  geom_point(aes(col = CI_lo < 0 & CI_hi > 0)) +
  geom_hline(yintercept = 0, color = "black", size = 0.5) +
  geom_errorbar(aes(ymin = CI_lo, ymax = CI_hi), width = 0.5, size = 0.2) +
  theme(
    text = element_text(size = 10),
    axis.text.x = element_text(size = 8),
    strip.text.x = element_text(size = 8),
    panel.background = element_rect(fill = "white", colour = "black"),
    legend.key.size = unit(3, "mm"),
    panel.grid = element_blank(),
    legend.position = "right",
    legend.text = element_text(size = 7),
    panel.grid.major.y = element_line(colour = "grey", size = 0.2, linetype = "dashed"),
    panel.grid.major.x = element_line(colour = "grey", size = 0.2, linetype = "dashed")
  ) +
  labs(y = "Beta (log OR)", x = "Annotations", col = "CI contains 0") +
  coord_flip() +
  facet_wrap(~trait, nrow = 1) +
  scale_color_manual(values = c("darkred", "gray"))

save_plot(filename = "../figures/2021-02-17_fgwas-t2d-enrich_beta.pdf", base_width = 10, base_height = 7)
```

```{r}
ggplot(fgwas_res, aes(fct_rev(annotation), trait)) +
  geom_tile(fill = "transparent") +
  geom_tile(data = fgwas_res %>% filter(!(CI_lo < 0 & CI_hi > 0)), aes(fill = estimate)) +
  geom_point(data = fgwas_res, shape = ifelse(fgwas_res$CI_lo < 0 & fgwas_res$CI_hi > 0, 32, 8)) +
  ggpubr::rotate_x_text(angle = 90) +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red") +
  theme(
    text = element_text(size = 10),
    axis.text.x = element_text(size = 8),
    strip.text.x = element_text(size = 8),
    panel.background = element_rect(fill = "white", colour = "black"),
    # legend.key.size=unit(3, "mm"),
    panel.grid = element_blank(),
    legend.position = "right",
    legend.text = element_text(size = 8),
    panel.grid.major.y = element_line(colour = "grey", size = 0.2, linetype = "dashed"),
    panel.grid.major.x = element_line(colour = "grey", size = 0.2, linetype = "dashed")
  ) +
  labs(x = "Annotations")
```


## AME / TF enrichment

See `2021-09-07_tf-enrichment.Rmd`


## Celltype-peak enrichment in modules

```{r}
gat_res_path <- glue("{root.dir}/work/07_module-peaks-celltype-enrichment/gat")
gat_res <- lapply(
   list.files(gat_res_path, "*"), function(celltype) {
     lapply(list.files(glue("{gat_res_path}/{celltype}"), '*'), function(mod_ver) {
         gat_files <- list.files(
           glue("{gat_res_path}/{celltype}/{mod_ver}"), "*.txt", full.names = T
         )
         
         do.call(rbind, lapply(gat_files, function(x) {
           df <- read.table(x, header = T)
           df$peak_set <- gsub(".gat.txt", "", basename(x))
           df$celltype <- celltype
           df$tss_def <- mod_ver
           df
         })) %>% bind_rows() 
     }) %>% bind_rows()
   }
) %>% bind_rows() %>%
  filter(grepl("sct", peak_set), tss_def == "hg19.tss.10kb_ext-distal")

gat_res$Pval <- as.numeric(str_match(gat_res$peak_set, ".*_(\\d+.\\d+)log2FC_(\\d?\\.?\\d*)_*")[, 2])
gat_res$log2FC <- as.numeric(str_match(gat_res$peak_set, ".*_(\\d+.\\d+)log2FC_(\\d?\\.?\\d*)_*")[, 3])
```

```{r}
plot_gat_heatmap <- function(df, celltype, title) {
  df %>% dplyr::filter(celltype == celltype) %>%
  mutate(zscore = ifelse(stddev > 0, (observed - expected) / stddev, 0)) %>%
  ggplot(., aes(gsub("Module-|.bed|.distal-10kb", "", annotation), track,
                fill = zscore, group = peak_set)) +
  geom_tile() +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red") +
  theme_custom(font_size = 12) +
  ggpubr::rotate_x_text() + 
  labs(x = "Modules", y = "Celltype-peaks", title = title)
}
```

```{r, fig.width = 12, fig.height = 2}
plot_gat_heatmap(gat_res %>% filter(grepl("alpha_sct", peak_set)),
                 celltype = "alpha-0.2", title = "Alpha cell modules (with Alpha peaks)") +
   facet_wrap(Pval ~ log2FC, scales = "free", ncol = 2)
```


```{r, fig.width = 12, fig.height = 4}
plot_gat_heatmap(gat_res %>% filter(grepl("beta_sct", peak_set)),
                 celltype = "beta-0.2", title = "Enrichment of celltype-specific distal peaks (Beta)") +
   facet_wrap(Pval ~ log2FC, scales = "free", ncol = 1)
```



## GAT footprint results

```{r}
gat_res_path <- glue("{root.dir}/control/module-peaks_gat/work/2021-06-09/gat")
gat_res <- lapply(
   list.files(gat_res_path, "*"), function(celltype) {
     lapply(list.files(glue("{gat_res_path}/{celltype}"), '*'), function(mod_ver) {
       lapply(list.files(glue("{gat_res_path}/{celltype}/{mod_ver}"), "*"), function(footprint_ver) {
         gat_files <- list.files(
           glue("{gat_res_path}/{celltype}/{mod_ver}/{footprint_ver}"), "*.txt", full.names = T
         )
         
         do.call(rbind, lapply(gat_files, function(x) {
           df <- read.table(x, header = T)
           df$motif <- gsub(".gat.txt", "", basename(x))
           df$footprint_ver <- footprint_ver
           df$celltype <- celltype
           df$mod_ver <- mod_ver
           df
         })) %>%
           mutate(annotation = gsub(".bed.gz", "", annotation)) %>%
           # mutate(annotation = factor(annotation, levels = unique(sort(as.numeric(annotation), decreasing = T)))) %>%
           mutate(zscore = (observed - expected) / stddev) %>%
           mutate(motif = gsub(".bound", "", motif)) %>%
           mutate(motif = gsub("Islet-consensus.", "", motif)) %>%
           mutate(motif = gsub(".Footprints", "", motif))
       }) %>% bind_rows()
     }) %>% bind_rows()
   }
) %>% bind_rows()
```


```{r}
gat_subset_df <- gat_res %>%
  filter(footprint_ver == "islet_meta_inspire" & celltype == "Beta.cell.RNA") %>%
  pivot_wider(id_cols = c(annotation), names_from = motif, values_from = zscore) %>%
  arrange(-as.numeric(annotation)) %>%
  column_to_rownames(var = "annotation") #%>%

ComplexHeatmap::Heatmap(gat_subset_df,  border = T,
                        column_names_gp =  gpar(fontsize = 10), row_names_gp = gpar(fontsize = 10),
                        name = "Beta", show_column_names =  F)

ComplexHeatmap::Heatmap(gat_subset_df,  border = T,
                        column_names_gp =  gpar(fontsize = 10), row_names_gp = gpar(fontsize = 10),
                        name = "Beta", show_column_names =  F, top_annotation = HeatmapAnnotation(names = anno_mark(at = which(grepl("RFX|PAX6", colnames(gat_subset_df))), labels = colnames(gat_subset_df)[which(grepl("RFX|PAX6", colnames(gat_subset_df)))])), column_dend_side = "bottom")
```


```{r}
gat_res %>%
  filter(footprint_ver == "islet_abcu196") %>%
  ggplot(., aes(annotation, motif, fill = zscore)) +
  geom_tile() +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red") +
  # geom_point(aes(shape = qvalue < 0.05)) +
  # scale_shape_manual(values = c(96, 8)) +
  facet_wrap(~ celltype, scales = "free_x") +
  theme_custom(font_size = 12) +
  ggpubr::rotate_x_text()
```

```{r}
gat_res %>%
  # filter(annotation_group == "hg19.tss.5u1d_ext") %>%
  mutate(zscore = (observed - expected) / stddev) %>%
  dplyr::select(annotation, zscore, motif, footprint_ver, annotation_group) %>%
  mutate(motif = gsub(".bound", "", motif)) %>%
  mutate(motif = gsub("Islet-consensus.", "", motif)) %>%
  mutate(motif = gsub(".Footprints", "", motif)) %>%
  pivot_wider(id_cols = c(annotation, motif, annotation_group), names_from = footprint_ver, values_from = zscore) %>%
  ggplot(., aes(islet_abcu196, islet_meta_inspire)) + 
  geom_point(alpha = .5) +
  ggrepel::geom_text_repel(aes(label = ifelse(abs(islet_abcu196 - islet_meta_inspire) > 2, motif, ""))) +
  geom_abline(aes(slope = 1, intercept = 0), col = 'red', linetype = 'dashed') +
  facet_wrap(~ annotation_group)
```

